---
title: "Regentrans Vignette"
author: "Sophie Hoffman"
date: "3/29/2021"
output: 
  html_document:
    toc: true
  
---

```{r,echo=FALSE, message=FALSE}
devtools::load_all()
```

```{r setup,echo=TRUE,message=FALSE}
library(tidyverse)
library(ape)
library(devtools)
library(ggtree)
library(pheatmap)
library(phytools)
library(gridExtra)
library(cowplot)
library(scales)

#set theme 
theme_set(theme_bw() + theme(strip.background = element_rect(fill="white",linetype='blank'), text=element_text(size=15)))
```


### Input data

```{r, message=FALSE}
#named vector of locations
locs <- metadata %>% select(isolate_id, facility) %>% deframe()
#named vector of patient IDs
pt <- metadata %>% select(isolate_id, patient_id) %>% deframe()
#named vector of isolate collection dates
dates <- metadata %>% select(isolate_id, collection_date) %>% deframe()
#fasta file 
fasta <- aln
#SNV distance matrix 
dists <- dists
#phylogenetic tree
tree <- tr
#patristic distances
pat_dists <- cophenetic.phylo(tree)
#patient transfer network
pt_flow <- pt_flow
#fsp
fsp <- fsp
```

### Subsetting to one isolate per patient per facility

```{r}
isolate_subset <- cbind(locs, pt) %>% unique() %>% rownames()
locs <- locs[isolate_subset]
pt <- pt[isolate_subset]
dates <- dates[isolate_subset]
dists <- dists[isolate_subset,isolate_subset]

```

# Choosing a pairwise SNV distance threshold

## Method 1: Use reference genome length and mutation rate

```{r}
ref_genome_length <- 5394056
mutation_rate <- 1.03e-6
round(2*ref_genome_length*mutation_rate)
```

## Method 2: Intra-facility pair fraction distribution

```{r}
# pre-calculated in vignette
# dists <- dist.dna(fasta, model = 'N', pairwise.deletion = TRUE, as.matrix = TRUE)
pair_types <- get_pair_types(dists = dists, locs = locs, pt = pt) #%>% subset_pairs()
pat_pair_types <- get_pair_types(dists = pat_dists, locs = locs, pt = pt) #%>% subset_pairs()

frac_intra <- get_frac_intra(pair_types = data.frame(pair_types %>% select(-pt1, -pt2)))
pat_pair_types_new <- pat_pair_types %>% 
  mutate(orig_dist = pairwise_dist, 
         orig_dist_bin = cut(pairwise_dist, 100), 
         pairwise_dist = as.numeric(cut(pairwise_dist, 100)))
pat_frac_intra <- pat_pair_types_new %>% select(-orig_dist_bin, -orig_dist, -pt1, -pt2) %>% data.frame() %>% 
  get_frac_intra() %>% left_join(pat_pair_types_new %>% select(pairwise_dist, orig_dist_bin) %>% unique())
```

```{r}
f2a <- frac_intra %>% 
  # filter(pairwise_dist <= 100) %>% 
  mutate(under = ifelse(pairwise_dist <= 10, '≤ 10 SNVs','Over threshold'), 
         under = ifelse(pairwise_dist <= 6, '≤ 6 SNVs',under),
         under = factor(under, levels = c('≤ 6 SNVs', '≤ 10 SNVs','Over threshold'))) %>% 
  ggplot(aes(x = pairwise_dist, y = frac_intra, fill = under)) + 
  geom_bar(stat = "identity", alpha = 0.5) + 
  scale_fill_manual(values = c('black','grey40','lightgrey')) + 
  labs(x = "Pairwise SNV distance", y = "Fraction of\nintra-facility pairs", fill = 'Possible distance\nthresholds') + 
  ylim(0, 1) #+ xlim(-1,100)


f2b <- pat_frac_intra %>% 
  # filter(orig_dist_bin %in% levels(pat_frac_intra$orig_dist_bin)[1:100]) %>% 
  mutate(under = ifelse(orig_dist_bin %in% levels(pat_frac_intra$orig_dist_bin)[1:4], 
                        'Branch length ≤ 2.56e-6\nmutations/site/year','Over threshold'),
  # under = ifelse(pairwise_dist <= 6, '≤ 6 SNVs',under),
  under = factor(under, levels = c('Branch length ≤ 2.56e-6\nmutations/site/year','Over threshold')),
  orig_dist_bin = as.numeric(gsub('\\(.*,|]','',orig_dist_bin))) %>%
  arrange(orig_dist_bin) %>% 
  ggplot(aes(x = orig_dist_bin, y = frac_intra, fill = under)) + 
  geom_bar(stat = "identity", alpha = 0.5) + 
  scale_fill_manual(values = c('black', 'lightgrey')) + 
  labs(x = "Pairwise patristic distance", y = "Fraction of\nintra-facility pairs", fill = 'Possible distance\nthreshold') +
  ylim(0, 1) #+
  # theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 0.5))

f2 <- plot_grid(plot_grid(f2a, NULL, rel_widths = c(1, 0.09), nrow = 1), 
                f2b, ncol = 1, labels = 'AUTO')

ggsave('manuscript_figures/Fig2.png', f2, width = 7, height = 5)
f2
```


# Is transmission occurring within and/or between locations? 

```{r, message=FALSE}
cols <- c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', '#000000')
names(cols) <- unique(locs)

tr_sub <- midpoint.root(ape::keep.tip(tr, names(locs)))
locs_tip <- c(locs[tr_sub$tip.label], rep(NA, Nnode(tr_sub)))
tr_facil <- ggtree(tr_sub) + geom_tippoint(aes(col = locs_tip)) + 
  scale_color_manual(values = cols) +
  labs(col = 'Facility') +
  theme(legend.position = 'none')
# tr_facil

clusters <- get_clusters(tr,locs, pureness = 1)
#dissect output
pure_subtree_info <- clusters$pure_subtree_info
subtrees <- clusters$subtrees

# cluster_plot <- ggplot(data = pure_subtree_info, aes(x = loc, y = subtr_size, color = loc)) + 
#   geom_jitter(position = position_jitter(width = 0.2, height = 0.1), alpha = 0.5,
#                        size = 2) + 
#   scale_color_manual(values = cols) +
#   labs(y = "#isolates in \nphylogenetic cluster", x = "", color = 'Facility') +
#   # ylim(c(0,12.5)) +
#   coord_flip()
# # cluster_plot 
```

```{r}
# get ids of isolates in each cluster
isolates_clusters <- sapply(1:nrow(clusters$pure_subtree_info), function(x){
  i <- clusters$pure_subtree_info$index[x]
  name <- clusters$pure_subtree_info$isolate_id[x]
  if(!is.na(i)){
    name <- clusters$subtrees[[i]]$tip.label
  }
  name
})

# get unique cluster names
facil_clusters <- sapply(unique(locs), function(x){
  clusts <- isolates_clusters[clusters$pure_subtree_info$loc == x]
  names(clusts) <- paste0(x, 1:length(clusts))
  cluster_nums <- unlist(sapply(names(clusts), function(x){
  rep(x, length(clusts[[x]]))
  }))
  names(cluster_nums) <- unlist(clusts)
  cluster_nums
}) %>% unname() %>% unlist()

head(facil_clusters)

day_diffs <- sapply(unique(facil_clusters), function(x){
  isolates <- names(facil_clusters)[facil_clusters == x]
  days <- sapply(isolates, function(i1){
    sapply(isolates, function(i2){
      b <- (days_between_isolates$isolate1 == i1 & days_between_isolates$isolate2 == i2 |
              days_between_isolates$isolate1 == i2 & days_between_isolates$isolate2 == i1)
      days_between_isolates$days_diff[b]
    })
  })
  max(unlist(days[lower.tri(days)]))
})

day_diffs[day_diffs == -Inf] = 0

cluster_plot <- bind_cols(cluster = names(day_diffs), 
          subtr_size = table(facil_clusters)[names(day_diffs)], 
          date_range = day_diffs) %>% 
  mutate(loc = gsub('[0-9]','',cluster), .before = 1) %>% 
  ggplot(aes(x = loc, y = subtr_size, color = loc, size = date_range)) + 
  geom_jitter(position = position_jitter(width = 0.2, height = 0.1), alpha = 0.5) + 
  scale_color_manual(values = cols) +
  labs(y = "# isolates in \nphylogenetic cluster", x = "", color = 'Facility', 
  size = 'Date range\n(days)') +
  # ylim(c(0,12.5)) +
  coord_flip()

f3 <- plot_grid(tr_facil, cluster_plot, labels = 'AUTO') #, rel_widths = c(0.4,0.6))

ggsave('manuscript_figures/Fig3.png', f3, width = 8, height = 8)
f3
```


```{r, message=FALSE}
pw_snv_plot <- pair_types %>% 
  ggplot(aes(x = pairwise_dist, fill = pair_type)) + 
  geom_histogram(position = 'identity', alpha = 0.4, bins = 30) +
  labs(x = "Pairwise SNV distance", y = "# pairs", fill = "") +
  geom_vline(xintercept = 10, col = 'darkgrey', size = 1) +
  theme(legend.justification = "top", 
        axis.line = element_line(colour = "black"),
        panel.border = element_blank())

pw_snv_plot_zoom <- pair_types %>% 
  filter(pairwise_dist <= 10) %>% 
  ggplot(aes(x = pairwise_dist, fill = pair_type)) + 
  geom_histogram(position = 'identity', alpha = 0.4, bins = 10) +
  labs(x = "Pairwise SNV\ndistance", y = "# pairs", fill = "") +
  theme(legend.position = 'none') +
  scale_x_continuous(breaks = pretty_breaks())

f4a <- pw_snv_plot + annotation_custom(ggplotGrob(pw_snv_plot_zoom), 
                                      xmin = 250, xmax = 450, ymin = 100, ymax = 2700)

# ggsave('manuscript_figures/Fig4.png', f4)
# f4

pat_pw_snv_plot <- pat_pair_types %>% 
  ggplot(aes(x = pairwise_dist, fill = pair_type)) + 
  geom_histogram(position = 'identity', alpha = 0.4, bins = 30) +
  labs(x = "Pairwise patristic distance", y = "# pairs", fill = "") +
  geom_vline(xintercept = 2.56e-06, col = 'darkgrey', size = 1) +
  theme(legend.justification = "top", 
        axis.line = element_line(colour = "black"),
        panel.border = element_blank())

pat_pw_snv_plot_zoom <- pat_pair_types %>%
  filter(pairwise_dist <= 2.56e-06) %>%
  ggplot(aes(x = pairwise_dist, fill = pair_type)) +
  geom_histogram(position = 'identity', alpha = 0.4, bins = 10) +
  labs(x = "Pairwise patristic\ndistance", y = "# pairs", fill = "") +
  theme(legend.position = 'none') #+
  # scale_x_continuous(breaks = pretty_breaks())

f4b <- pat_pw_snv_plot + annotation_custom(ggplotGrob(pat_pw_snv_plot_zoom), 
                                    xmin = 5.5e-5, xmax = 9.5e-5, ymin = 50, ymax = 2650)

f4 <- plot_grid(f4a, f4b, ncol = 1, labels = 'AUTO')
ggsave('manuscript_figures/Fig4.png', f4, height = 7, width = 6)
f4

```

```{r}
left_join(pair_types, days_between_isolates) %>% 
  ggplot(aes(x = days_diff, y = pairwise_dist)) +
  geom_bin_2d() +
  geom_smooth(method = 'lm', color = 'white') +
  labs(x = '# days between samples', y = 'Pairwise SNV distance', fill = '# pairs')

left_join(pat_pair_types, days_between_isolates) %>% 
  ggplot(aes(x = days_diff, y = pairwise_dist)) +
  geom_bin_2d() +
  geom_smooth(method = 'lm', color = 'white') +
  labs(x = '# days between samples', y = 'Pairwise patristic distance', fill = '# pairs')
```


# What locations is transmission occurring within/between?

### Fsp

Fsp is a measure of similarity between populations at different facilities. Value of 0 indicates that the populations are identical, while value of 1 indicates that the populations are completely different. 

```{r, eval=FALSE, message=FALSE}
# fsp <- get_facility_fsp(fasta, locs)
fsp <- regentrans::Fsp
pheatmap(fsp, border_color = 'white',
         legend_breaks = c(0, 0.2, 0.4, 0.6, 0.8, max(fsp)), 
         legend_labels = c(0, 0.2, 0.4, 0.6, 0.8, 'Fsp'), 
         filename = 'manuscript_figures/Fig5.png', width = 7, height = 6)

pheatmap(fsp, border_color = 'white',
         legend_breaks = c(0, 0.2, 0.4, 0.6, 0.8, max(fsp)), 
         legend_labels = c(0, 0.2, 0.4, 0.6, 0.8, 'Fsp'))
```

### SNV distances 

```{r, message=FALSE, warning=FALSE}
# get snv summary
snv_summary <- summarize_pairs(pair_types %>% select(-pt1,-pt2) %>% data.frame(),summary_fns = c('min'), threshs = c(6, 10)) 

# get facility summary
# facil_summary <- snv_summary %>% 
#     mutate(pair_type = factor(ifelse(loc1 == loc2, 'Intra-facility pair', 'Inter-facility pair'), 
#                             levels = c('Intra-facility pair', 'Inter-facility pair'))) %>% 
#   group_by(loc1, pair_type, leq_6, leq_10) %>% summarize() %>% 
#   pivot_longer(cols = c(leq_6, leq_10)) %>% rename(thresh = name, n_pairs = value) %>% 
#   mutate(thresh = paste(gsub('leq_','≤ ', thresh), 'SNVs')) 
facil_summary <- snv_summary %>% 
  mutate(pair_type = factor(ifelse(loc1 == loc2, 'Intra-facility pair', 'Inter-facility pair'), levels = c('Intra-facility pair', 'Inter-facility pair'))) %>% 
  group_by(loc1, pair_type) %>% summarize(leq_6 = sum(leq_6), leq_10 = sum(leq_10)) %>%
  pivot_longer(cols = c(leq_6, leq_10)) %>% rename(thresh = name, n_pairs = value) %>%
  mutate(thresh = paste(gsub('leq_','≤ ', thresh), 'SNVs')) 

# get locations to keep (at least 1 closely related pair)
keep_locs <- facil_summary %>% group_by(loc1) %>% summarize(sum_all = sum(n_pairs)) %>% filter(sum_all != 0) %>% select(loc1) %>% deframe()

# plot
f6 <- facil_summary %>% filter(loc1 %in% keep_locs) %>% 
  tidyr::complete(loc1, pair_type, thresh) %>% 
  ggplot(aes(x = reorder(loc1, -n_pairs, sum), y = n_pairs, fill = pair_type)) +
  geom_col(position = 'dodge') +
       scale_fill_manual(values=c("cadetblue", "salmon")) + 
         labs(y = "# closely related isolate pairs", x = "Facility", fill = '', color = '') +
  facet_grid(factor(thresh, levels = c('≤ 6 SNVs','≤ 10 SNVs'))~.)


ggsave('manuscript_figures/Fig6.png',f6, width = 7, height = 5)
f6
```

```{r}
time1 <- pair_types %>% 
  mutate(date1 = dates[isolate1], date2 = dates[isolate2]) %>% 
  filter(date1 == date2) %>% 
  ggplot(aes(x = pairwise_dist, fill = pair_type)) + 
  geom_histogram(position = 'identity', alpha = 0.4, bins = 30) +
  labs(x = "Pairwise SNV distance", y = "# pairs", fill = "") +
  facet_grid(~date1) + 
  theme(legend.justification = "top", 
        axis.line = element_line(colour = "black"),
        panel.border = element_blank())

time2 <- pair_types %>% 
  mutate(date1 = dates[isolate1], date2 = dates[isolate2]) %>% 
  filter(date1 == date2) %>% 
  ggplot(aes(x = pairwise_dist, fill = pair_type)) + 
  geom_histogram(position = position_fill(), alpha = 0.4, bins = 30) +
  # xlim(0,20) +
  labs(x = "Pairwise SNV distance", y = "Proportion of pairs", fill = "") +
  facet_grid(~date1) + 
  theme(legend.justification = "top", 
        axis.line = element_line(colour = "black"),
        panel.border = element_blank())

f7 <- plot_grid(time1, time2, nrow = 2, labels = 'AUTO')
ggsave('manuscript_figures/Fig7.png', f8, width = 7, height = 5)
f7
```

```{r}
pair_info <- merge_inter_summaries(
  make_long_form(pt_flow, col_names = c('loc1','loc2','pt_flow')), 
  snv_summary, 
  make_long_form(fsp)) %>% 
  filter(loc1 != loc2)

# fsp_sub <- fsp[rownames(pt_flow), colnames(pt_flow)]

# fsp_flow <- full_join(reshape2::melt(as.matrix(pt_flow)) %>% rename(pt_flow = value),
          # reshape2::melt(as.matrix(fsp_sub)) %>% rename(fsp = value)) %>% rename(loc1 = Var1, loc2 = Var2)

# thresh = 6

# snv_dist_summary <- pair_types %>% dplyr::group_by(loc1, loc2) %>% dplyr::summarize(n_closely_related_pairs = sum(pairwise_dist <= thresh)) %>% dplyr::filter(loc1 != loc2)

# pair_info <- full_join(fsp_flow, snv_dist_summary)

flow_fsp_spear <- cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$fsp)],
                                 pair_info$fsp[!is.na(pair_info$pt_flow) & !is.na(pair_info$fsp)], 
                                 method = 'spearman')

flow_fsp_plot <- pair_info %>% 
  filter(!is.na(pt_flow)) %>% 
  ggplot(aes(x=pt_flow,y=fsp)) +
  geom_point() + geom_smooth(method='lm') + scale_x_log10() +
  labs(x = 'Patient flow', y = 'Fsp') + 
  annotate('text', x = 0.004, y = 0.7, 
           label = paste(paste('Spearman rho =', round(flow_fsp_spear$estimate, 2)),'\n',
                         paste('Spearman p =', signif(flow_fsp_spear$p.value, 1))))

flow_snv_spear <- cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$leq_10)],
                                 pair_info$leq_10[!is.na(pair_info$pt_flow) & !is.na(pair_info$leq_10)], 
                                 method = 'spearman')

flow_snv_plot <- pair_info %>% 
  filter(!is.na(pt_flow)) %>% 
  ggplot(aes(x=pt_flow,y=leq_10)) +
  geom_point() + geom_smooth(method='lm') + scale_x_log10() +
  labs(x = 'Patient flow', y = '# closely related\npairs (≤ 10 SNVs)') + 
  annotate('text', x = 0.004, y = 6.5, 
           label = paste(paste('Spearman rho =', round(flow_snv_spear$estimate, 2)),'\n',
                         paste('Spearman p =', signif(flow_snv_spear$p.value, 1)))) #+ ylim(c(0,100))

flow_min_snv_spear <- cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$dist_min)],
                                 pair_info$dist_min[!is.na(pair_info$pt_flow) & !is.na(pair_info$dist_min)], 
                                 method = 'spearman')

flow_min_snv_plot <- pair_info %>%
  filter(!is.na(pt_flow)) %>% 
  ggplot(aes(x=pt_flow,y=dist_min)) +
  geom_point() + geom_smooth(method='lm') + scale_x_log10() +
  labs(x = 'Patient flow', y = 'Minimum pairwise \nSNV distance') +
  annotate('text', x = 0.004, y = 30, 
           label = paste(paste('Spearman rho =', round(flow_min_snv_spear$estimate, 2)),'\n',
                         paste('Spearman p =', signif(flow_min_snv_spear$p.value, 1)))) #+ ylim(c(0,40))

f8 <- plot_grid(flow_fsp_plot, flow_min_snv_plot, flow_snv_plot, ncol = 1, labels = 'AUTO')
ggsave('manuscript_figures/Fig7.png',f8, width = 6, height = 8)
f8

#f7 <- plot_grid(flow_fsp_plot, flow_snv_plot, ncol = 1, labels = 'AUTO')
#ggsave('manuscript_figures/Fig7.png',f7)
#f7
```

```{r, eval = FALSE}
cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$fsp)],
                                 pair_info$fsp[!is.na(pair_info$pt_flow) &
                                                 !is.na(pair_info$fsp)],  
         method = 'spearman')

cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$fsp) &
                             pair_info$pt_flow < 2e-3],
                                 pair_info$fsp[!is.na(pair_info$pt_flow) &
                                                 !is.na(pair_info$fsp) &
                             pair_info$pt_flow < 2e-3],  
         method = 'spearman')

cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$leq_10)],
                                 pair_info$leq_10[!is.na(pair_info$pt_flow) & !is.na(pair_info$leq_10)], 
                                 method = 'spearman')

cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$leq_10) &
                             pair_info$leq_10 < 8],
                                 pair_info$leq_10[!is.na(pair_info$pt_flow) & !is.na(pair_info$leq_10) &
                             pair_info$leq_10 < 8], 
                                 method = 'spearman')

cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$dist_min)],
                                 pair_info$dist_min[!is.na(pair_info$pt_flow) & !is.na(pair_info$dist_min)], 
                                 method = 'spearman')

cor.test(pair_info$pt_flow[!is.na(pair_info$pt_flow) & !is.na(pair_info$dist_min) &
                             pair_info$dist_min < 20],
                                 pair_info$dist_min[!is.na(pair_info$pt_flow) & !is.na(pair_info$dist_min)&
                             pair_info$dist_min < 20], 
                                 method = 'spearman')
```


```{r}
# read_tsv('~/Desktop/isolate_info.tsv') %>% select(isolate_no, clades) %>% mutate(date = dates[isolate_no]) %>% select(clades, date) %>% 
#   table() %>% data.frame() %>% pivot_wider(names_from = date, values_from = freq) %>% 
#   mutate(frac_2015 = `2015`/(`2014`+`2015`))
```


```{r}
# these are deidentified
longs <- c(A = NA, B = NA, C = -930605.082601, D = -930605.009738, E = -930604.936812, 
F = -930604.912683, G = -930604.930899, H = -930605.236274, I = -930605.065595, 
J = NA, K = NA, L = NA, M = -930605.109771, N = NA, O = -930605.187588, 
P = -930604.896185, Q = NA, R = NA, S = NA, T = -930605.237532, 
U = NA)
lats <- c(A = NA, B = NA, C = -740730.013261, D = -740730.378523, E = -740729.967478, 
F = -740729.917235, G = -740729.643727, H = -740729.995281, I = -740729.894068, 
J = NA, K = NA, L = NA, M = -740730.297362, N = NA, O = -740729.22271, 
P = -740729.566327, Q = NA, R = NA, S = NA, T = -740729.891502, 
U = NA)

facil_coord <- bind_cols(facil=names(longs), long=longs, lat=lats)

loc_n <- locs %>% table() %>% as.data.frame() %>% `colnames<-`(c("facil", "n"))

facil_geo <- facil_coord %>% left_join(loc_n)

# bind_cols(facil=names(longs), long=longs, lat=lats) %>% ggplot(aes(x = lats, y = longs)) + geom_point() + theme_map()

snv_geo <- pair_types %>% dplyr::group_by(loc1, loc2) %>% dplyr::summarize(n_closely_related_pairs = sum(pairwise_dist <= 7)) %>% dplyr::filter(loc1 != loc2, n_closely_related_pairs > 0) %>%
  left_join(facil_geo, by = c("loc1" = "facil")) %>% rename(facil_1_lat = lat, facil_1_long = long, facil_1_n = n) %>% left_join(facil_geo, by = c("loc2" = "facil")) %>% rename(facil_2_lat = lat, facil_2_long = long, facil_2_n = n)

geo_map <- facil_geo %>% ggplot(aes(x=long, y=lat)) +
  geom_curve(aes(x = facil_1_long, y = facil_1_lat, xend = facil_2_long, yend = facil_2_lat,
                 col = as.character(n_closely_related_pairs), size = n_closely_related_pairs/10),
             data = snv_geo, curvature = 0.33) +
  geom_point(aes(size=n, fill=n), shape = 21) + #, color = "lightblue") + 
  # scale_color_gradient2(low = adjustcolor('lightgrey', alpha.f = 0.75), high = adjustcolor('black', alpha.f = 0.75)) +
  scale_color_manual(values = gray.colors(7, start = 0.3, end = 0.9, gamma = 2.2, alpha = 0.75, rev = TRUE)) +
  scale_fill_viridis_c(option = 'viridis', direction = -1) +
  scale_size_continuous(range = c(1, 12), guide = FALSE) +
  theme_map() + #theme(legend.position = "none") 
  labs(size = '# samples', fill = '# samples', col = '# closely related\npairs (≤ 6 SNVs)')

geo_dat <- left_join(pair_info, facil_geo, by = c('loc1' = 'facil')) %>% rename(facil_1_long = long, facil_1_lat = lat) %>% 
  select(-n) %>% left_join(., facil_geo, by = c('loc2' = 'facil')) %>% rename(facil_2_long = long, facil_2_lat = lat) %>% 
  select(-n) %>% 
  mutate(geo_dist = sqrt((facil_1_long-facil_2_long)^2+(facil_1_lat-facil_2_lat)^2))

geo_fsp_spear <- cor.test(geo_dat$geo_dist[!is.na(geo_dat$geo_dist) & !is.na(geo_dat$fsp)],
                                 geo_dat$fsp[!is.na(geo_dat$geo_dist) & !is.na(geo_dat$fsp)], 
                                 method = 'spearman')

geo_min_snv_spear <- cor.test(geo_dat$geo_dist[!is.na(geo_dat$geo_dist) & !is.na(geo_dat$dist_min)],
                                 geo_dat$dist_min[!is.na(geo_dat$geo_dist) & !is.na(geo_dat$dist_min)], 
                                 method = 'spearman')

geo_snv_spear <- cor.test(geo_dat$geo_dist[!is.na(geo_dat$geo_dist) & !is.na(geo_dat$leq_10)],
                                 geo_dat$leq_10[!is.na(geo_dat$geo_dist) & !is.na(geo_dat$leq_10)], 
                                 method = 'spearman')

geo_fsp <- geo_dat %>% 
  filter(!is.na(geo_dist)) %>% 
  ggplot(aes(x = geo_dist, y = fsp)) +
  geom_point() + geom_smooth(method = 'lm') +
  labs(x = 'Physical distance between facilities', y = 'Fsp') + 
  annotate('text', x = 0.9, y = 0.3, 
           label = paste0(paste('Spearman rho =', round(geo_fsp_spear$estimate, 2)),'\n',
                         paste('Spearman p =', signif(geo_fsp_spear$p.value, 1))))

geo_min_snv <- geo_dat %>% 
  filter(!is.na(geo_dist)) %>% 
  ggplot(aes(x = geo_dist, y = dist_min)) + geom_point() + geom_smooth(method = 'lm') +
  labs(x = 'Physical distance between facilities', y = 'Minimum pairwise \nSNV distance') + 
  annotate('text', x = 0.9, y = 30, 
           label = paste0(paste0('Spearman rho = ', round(geo_min_snv_spear$estimate, 2)),'\n',
                         paste0('Spearman p = ', signif(geo_min_snv_spear$p.value, 1))))

geo_snv <- geo_dat %>% 
  filter(!is.na(geo_dist)) %>% 
  ggplot(aes(x = geo_dist, y = leq_10)) + geom_point() + geom_smooth(method = 'lm') +
  labs(x = 'Physical distance between facilities', y = '# closely related pairs\n(≤ 10 SNVs)') + 
  annotate('text', x = 0.9, y = 10, 
           label = paste0(paste0('Spearman rho = ', round(geo_snv_spear$estimate, 2)),'\n',
                         paste0('Spearman p = ', signif(geo_snv_spear$p.value, 1))))



f9 <- plot_grid(plot_grid(NULL, geo_map, NULL, nrow = 1, rel_widths = c(0.1,0.9,0), labels = c('','A','')), 
                plot_grid(geo_fsp, geo_min_snv, geo_snv, labels = c('B','C','D'), nrow = 1), 
                ncol = 1, rel_heights = c(0.6, 0.4))
ggsave('manuscript_figures/Fig9.png', f9, width = 12, height = 12)
f9
```




