<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A framework and R package for using genomics to study regional pathogen transmission • regentrans</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A framework and R package for using genomics to study regional pathogen transmission">
<meta property="og:description" content="Tools and framework for studying regional pathogen transmission using genomic data.
    See the GitHub page &lt;https://github.com/Snitkin-Lab-Umich/regentrans&gt; 
    for more information, documentation, and examples.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">regentrans</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Introduction.html">Introduction</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Snitkin-Lab-Umich/regentrans/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="regentrans-r-package-for-investigating-regional-pathogen-transmission-using-genomic-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#regentrans-r-package-for-investigating-regional-pathogen-transmission-using-genomic-data" class="anchor"></a>regentrans: R package for investigating regional pathogen transmission using genomic data</h1></div>
<p><img src="reference/figures/regentrans_logo.png" width="200"></p>
<p>Website: <a href="https://snitkin-lab-umich.github.io/regentrans/" class="uri">https://snitkin-lab-umich.github.io/regentrans/</a></p>
<p>Preprint: <a href="https://www.medrxiv.org/content/10.1101/2021.07.25.21261097v1" class="uri">https://www.medrxiv.org/content/10.1101/2021.07.25.21261097v1</a></p>
<p><code>regentrans</code> can be installed using the command <code><a href="https://devtools.r-lib.org//reference/remote-reexports.html">devtools::install_github('Snitkin-Lab-Umich/regentrans')</a></code> in R</p>
<p>Examples of how to interrogate all of the questions in the table below can be found in the <a href="https://snitkin-lab-umich.github.io/regentrans/articles/Introduction.html">introductory vignette</a>.</p>
<p>Questions regentrans can help investigate:</p>
<table class="table">
<colgroup>
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th>Question</th>
<th>Method</th>
<th>regentrans function(s)</th>
<th>Required Data</th>
<th>Optional Data</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>How do you choose pairwise SNV distance thresholds?</td>
<td>Fraction of intra-facility pairs for different pairwise SNV distances</td>
<td><code><a href="reference/get_frac_intra.html">get_frac_intra()</a></code></td>
<td>output of <code><a href="reference/get_pair_types.html">get_pair_types()</a></code>
</td>
<td></td>
</tr>
<tr class="even">
<td>Is transmission occurring within and/or between locations?</td>
<td>Phylogenetic clustering of isolates from the same location</td>
<td><code><a href="reference/get_clusters.html">get_clusters()</a></code></td>
<td>Phylogenetic tree, isolate location information</td>
<td></td>
</tr>
<tr class="odd">
<td></td>
<td>Pairwise SNV distances within and between facilities</td>
<td><code><a href="reference/get_pair_types.html">get_pair_types()</a></code></td>
<td>A pairwise SNV distance matrix (can be created using <code><a href="https://rdrr.io/pkg/ape/man/dist.dna.html">ape::dist.dna()</a></code> on fasta file of variants), isolate location information</td>
<td></td>
</tr>
<tr class="even">
<td>What locations is transmission occurring within/between?</td>
<td>Population-level similarity between locations (Fsp)</td>
<td><code><a href="reference/get_facility_fsp.html">get_facility_fsp()</a></code></td>
<td>Fasta file of variants, isolate location information</td>
<td></td>
</tr>
<tr class="odd">
<td></td>
<td>Number of closely related pairs within and between facilities</td>
<td>Plot code in vignette</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Have transmission dynamics changed over time?</td>
<td>Methods above but split over time</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>Is transmission occurring along paths of higher patient/person flow?</td>
<td>Compare patient/person flow between locations to inter-location pairwise SNV distances or Fsp</td>
<td><code><a href="reference/get_patient_flow.html">get_patient_flow()</a></code></td>
<td>Patient transfer network edge matrix</td>
<td></td>
</tr>
<tr class="even">
<td></td>
<td>Compare patient/person flow between locations to populatoin-level similarity between locations</td>
<td>
<code><a href="reference/get_patient_flow.html">get_patient_flow()</a></code>, <code><a href="reference/get_facility_fsp.html">get_facility_fsp()</a></code>
</td>
<td>Patient transfer network, fasta file of variants, isolate location information</td>
<td></td>
</tr>
<tr class="odd">
<td>Are there any observable geographic trends in prevalence/transmission?</td>
<td>Visualize prevalence and closely related pairs or Fsp</td>
<td>Plot code in vignette</td>
<td>Geographic locations of each facility, patient transfer network, summary metric of genomic similarity</td>
<td></td>
</tr>
</tbody>
</table>
<div id="questions" class="section level2">
<h2 class="hasAnchor">
<a href="#questions" class="anchor"></a>Questions</h2>
<p>Please open an <a href="https://github.com/Snitkin-Lab-Umich/regentrans/issues">issue</a> or contact Evan Snitkin (<a href="mailto:esnitkin@umich.edu" class="email">esnitkin@umich.edu</a>) with any questions.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/Snitkin-Lab-Umich/regentrans/">https://​github.com/​Snitkin-Lab-Umich/​regentrans/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/Snitkin-Lab-Umich/regentrans/issues">https://​github.com/​Snitkin-Lab-Umich/​regentrans/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Sophie Hoffman <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-2518-6422" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Zena Lapp <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4674-2176" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Joyce Wang <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-8674-1015" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Evan Snitkin <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-8409-278X" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Sophie Hoffman, Zena Lapp, Joyce Wang, Evan Snitkin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
