<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction • regentrans</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction">
<meta property="og:description" content="regentrans">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">regentrans</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">Introduction</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Snitkin-Lab-Umich/regentrans/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Snitkin-Lab-Umich/regentrans/blob/master/vignettes/Introduction.Rmd"><code>vignettes/Introduction.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction.Rmd</code></div>

    </div>

    
    
<p>The goal of <code>regentrans</code> is to give users a framework and corresponding package for using genomics to study regional pathogen transmission.</p>
<p>To install the package from GitHub, use the command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install devtools if you don't already have it installed</span>
<span class="co"># install.packages('devtools') </span>

<span class="co"># install regentrans</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">'Snitkin-Lab-Umich/regentrans'</span><span class="op">)</span>

<span class="co"># load regentrans</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://snitkin-lab-umich.github.io/regentrans/index.html">regentrans</a></span><span class="op">)</span></code></pre></div>
<p>Other packages you’ll need to load to run this tutorial include:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/">ape</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/">devtools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/treedata-book/">ggtree</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/liamrevell/phytools">phytools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span>

<span class="co"># set theme for plots </span>
<span class="fu">theme_set</span><span class="op">(</span><span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>strip.background <span class="op">=</span> <span class="fu">element_rect</span><span class="op">(</span>fill<span class="op">=</span><span class="st">"white"</span>,linetype<span class="op">=</span><span class="st">'blank'</span><span class="op">)</span>, text<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="a-note-on-population-genomics" class="section level1">
<h1 class="hasAnchor">
<a href="#a-note-on-population-genomics" class="anchor"></a>A note on population genomics</h1>
<p>It is outside the scope of this framework and corresponding package to discuss population genomic analysis more generally. We direct you to this review for more information on what to consider when developing and executing a population genomics study: <a href="https://pubmed.ncbi.nlm.nih.gov/28513284/">Best Practice for Population Genetic Analysis</a>.</p>
</div>
<div id="data-prep" class="section level1">
<h1 class="hasAnchor">
<a href="#data-prep" class="anchor"></a>Data prep</h1>
<p>In general, the data pre-processing pipeline is described in this figure:</p>
<div class="figure">
<img src="manuscript_figures/Fig1.png" alt="regentrans pre-processing workflow." style="width:75.0%"><p class="caption"><code>regentrans</code> pre-processing workflow.</p>
</div>
<p>More specifically, to use <code>regentrans</code>, you will need your data in a certain format. It is important to ensure that you have all relevant information for all of the isolates of interest for your analysis. Each isolate should be represented in the metadata as well as in the fasta file of variants and phylogenetic tree. <code>regentrans</code> functions will subset the data for you to include only ones present in all of the required function inputs.</p>
<p>First, we provide examples of how you can read in your data. Then, we describe the built-in data that we use here in the vignette and that you can use to test out the package.</p>
<div id="how-to-read-in-your-own-data" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-read-in-your-own-data" class="anchor"></a>How to read in your own data</h2>
<p>Note that this code will not run successfully unless you change the paths to be correct.</p>
<div id="metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata" class="anchor"></a>Metadata</h3>
<p>The only required metadata is location, but additional information can also be incorporated into the analysis including patient, sequence type, and collection date.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this is if your metadata is in a csv file</span>
<span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"/path/to/metadata.csv"</span><span class="op">)</span></code></pre></div>
</div>
<div id="genomic-data" class="section level3">
<h3 class="hasAnchor">
<a href="#genomic-data" class="anchor"></a>Genomic data</h3>
<p>Either a variant alignment or a phylogenetic tree are required for most functions in <code>regentrans</code>.</p>
<div id="variant-alignment" class="section level4">
<h4 class="hasAnchor">
<a href="#variant-alignment" class="anchor"></a>Variant alignment</h4>
<p>The variant alignment can be read in to R using the following command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this is if your alignment is in a fasta file</span>
<span class="va">aln</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.dna.html">read.dna</a></span><span class="op">(</span><span class="st">"/path/to/aln.fasta"</span>,
                     format <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></code></pre></div>
</div>
<div id="pairwise-snv-distance-matrix" class="section level4">
<h4 class="hasAnchor">
<a href="#pairwise-snv-distance-matrix" class="anchor"></a>Pairwise SNV distance matrix</h4>
<p>Some <code>regentrans</code> functions require a genomic distance matrix as input. A pairwise single nucleotide variant (SNV) distance matrix can be created using the fasta file:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dists</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/dist.dna.html">dist.dna</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">aln</span>, <span class="co"># DNAbin object as read in above</span>
                       as.matrix <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># return as matrix</span>
                       model <span class="op">=</span> <span class="st">"N"</span>, <span class="co"># count pairwise distances</span>
                       pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># delete sites with missing data in a pairwise way</span>
                       <span class="op">)</span></code></pre></div>
</div>
<div id="phylogenetic-tree" class="section level4">
<h4 class="hasAnchor">
<a href="#phylogenetic-tree" class="anchor"></a>Phylogenetic tree</h4>
<p>We recommend using <a href="http://www.iqtree.org/">IQ-TREE</a> to generate a maximum likelihood phylogenetic tree for your dataset. However, there are also ways to make simple phylogenies in R using a variant alignment.</p>
<p>A phylogenetic tree in Newick format can be read into R using the following command:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this is if the tree is in Newick format</span>
<span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="st">"path/to/data.treefile"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="patient-transfer-network" class="section level3">
<h3 class="hasAnchor">
<a href="#patient-transfer-network" class="anchor"></a>Patient transfer network</h3>
<p>A patient transfer network representing facilities that isolates came from represented by a data.frame with three columns: <code>source_facil</code>, <code>dest_facil</code> and <code>n_transfers</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this is if your patient transfer network is in a csv file</span>
<span class="va">pt_trans_df</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="st">"/path/to/pt_trans_net.csv"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="input-data-format-using-the-built-in-regentrans-data" class="section level2">
<h2 class="hasAnchor">
<a href="#input-data-format-using-the-built-in-regentrans-data" class="anchor"></a>Input data format (&amp; using the built-in <code>regentrans</code> data)</h2>
<p>The example data from the package uses carbapenem-resistant <em>Klebsiella pneumoniae</em> (CRKP) sequence type (ST) 258 whole-genome sequencing data from <a href="https://aac.asm.org/content/63/11/e01622-19">Han et al. 2019</a>. This data is a comprehensive collection of clinical CRKP isolates collected from long-term acute care hospitals across the U.S. over the course of a year. The data was processed as in <a href="https://www.medrxiv.org/content/10.1101/2021.06.11.21258758v1">Lapp et al. 2021</a>.</p>
<p>If you want to use your own data as input to the functions in the vignette, format your data according to the instructions above and set the following objects equal to your data. If you will not be using any of the optional data, make sure to assign those objects to <code>NULL</code>.</p>
<div id="metadata-1" class="section level3">
<h3 class="hasAnchor">
<a href="#metadata-1" class="anchor"></a>Metadata</h3>
<p>The <code>regentrans</code> metadata is in the <code>metadata</code> object. We have information on the sample ID, patient ID, facility, and collection date:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></code></pre></div>
<pre><code>##   isolate_id patient_id facility collection_date
## 1    PCMP_H1         p1        A            2014
## 2    PCMP_H2         p2        B            2014
## 3    PCMP_H3         p3        C            2014
## 4    PCMP_H4         p4        D            2014
## 5    PCMP_H5         p5        E            2014
## 6  PCMP_H223         p5        F            2015</code></pre>
</div>
<div id="variant-alignment-1" class="section level3">
<h3 class="hasAnchor">
<a href="#variant-alignment-1" class="anchor"></a>Variant alignment</h3>
<p>The variant alignment is a DNAbin object that we can use to calculate the pairwise SNV distances between our samples:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "DNAbin"</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">aln</span></code></pre></div>
<pre><code>## 413 DNA sequences in binary format stored in a matrix.
## 
## All sequences of same length: 83976 
## 
## Labels:
## PCMP_H100
## PCMP_H101
## PCMP_H103
## PCMP_H104
## PCMP_H105
## PCMP_H106
## ...
## 
## More than 10 million bases: not printing base composition.
## (Total: 34.68 Mb)</code></pre>
</div>
<div id="pairwise-snv-distance-matrix-1" class="section level3">
<h3 class="hasAnchor">
<a href="#pairwise-snv-distance-matrix-1" class="anchor"></a>Pairwise SNV distance matrix</h3>
<p>We have provided the pairwise SNV distance matrix as part of the package as it is a bit timely to compute, but it can also easily be generated using the alignment we provideusing the command described above. Here is a subset of the distance matrix:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dists</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code>##           PCMP_H100 PCMP_H101 PCMP_H103 PCMP_H104 PCMP_H105
## PCMP_H100         0       148       194        69       171
## PCMP_H101       148         0       128       160        47
## PCMP_H103       194       128         0       208       156
## PCMP_H104        69       160       208         0       184
## PCMP_H105       171        47       156       184         0</code></pre>
<p>The row and column names are the names of the samples and the values in the matrix are the pairwise SNV distances between each isolate pair. Note that this is a symmetric matrix.</p>
</div>
<div id="phylogenetic-tree-1" class="section level3">
<h3 class="hasAnchor">
<a href="#phylogenetic-tree-1" class="anchor"></a>Phylogenetic tree</h3>
<p>We also provide a maximum likelihood phylogenetic tree that was generated using <a href="http://www.iqtree.org/">IQ-TREE</a>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tr</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 413 tips and 412 internal nodes.
## 
## Tip labels:
##   PCMP_H100, PCMP_H98, PCMP_H123, PCMP_H141, PCMP_H376, PCMP_H384, ...
## Node labels:
##   100, 99, 99, 100, 83, 100, ...
## 
## Rooted; includes branch lengths.</code></pre>
</div>
<div id="patient-transfer-network-1" class="section level3">
<h3 class="hasAnchor">
<a href="#patient-transfer-network-1" class="anchor"></a>Patient transfer network</h3>
<p>While we are not able to provide the true patient transfer network related to the example dataset, we have generated a simulated patient transfer network that we have included here:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pt_trans_df</span><span class="op">)</span></code></pre></div>
<pre><code>##   source_facil dest_facil n_transfers
## 1            H          D          17
## 2            G          D          14
## 3            C          D         474
## 4            T          D         182
## 5            F          D          60
## 6            I          D          63</code></pre>
<p>This data frame contains three columns:</p>
<ol style="list-style-type: decimal">
<li>
<code>source_facil</code>: The source facility.</li>
<li>
<code>dest_facil</code>: The destination facility.</li>
<li>
<code>n_transfers</code>: The number of patient transfers between source and destination facility over some length of time (e.g. 1 day, 1 month, or 1 year).</li>
</ol>
</div>
</div>
<div id="subsetting-to-one-isolate-per-patient-per-facility" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-to-one-isolate-per-patient-per-facility" class="anchor"></a>Subsetting to one isolate per patient per facility</h2>
<p>As we have multiple samples from some patients, we first subset to the first isolate from each patient per facility to avoid overestimating transmission. We can do this in R as follows:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># subset metadata to include the first isolate from each patient per facility</span>
<span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op">%&gt;%</span> <span class="fu">arrange</span><span class="op">(</span><span class="va">collection_date</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">patient_id</span>, <span class="va">facility</span><span class="op">)</span><span class="op">)</span>

<span class="co"># isolate subset</span>
<span class="va">isolate_subset</span> <span class="op">&lt;-</span> <span class="va">metadata</span><span class="op">$</span><span class="va">isolate_id</span>

<span class="co"># subset alignment</span>
<span class="va">aln</span> <span class="op">&lt;-</span> <span class="va">aln</span><span class="op">[</span><span class="va">isolate_subset</span>,<span class="op">]</span>

<span class="co"># subset tree</span>
<span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/drop.tip.html">keep.tip</a></span><span class="op">(</span><span class="va">tr</span>, <span class="va">isolate_subset</span><span class="op">)</span>

<span class="co"># subset distance matrix</span>
<span class="va">dists</span> <span class="op">&lt;-</span> <span class="va">dists</span><span class="op">[</span><span class="va">isolate_subset</span>, <span class="va">isolate_subset</span><span class="op">]</span></code></pre></div>
</div>
<div id="extracting-location-as-a-vector" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-location-as-a-vector" class="anchor"></a>Extracting location as a vector</h2>
<p>For many of the <code>regentrans</code> functions, the required location input is a named vector where the elements are locations (e.g. facility) and the names are the sample ID. To get the location information from the built-in metadata, we can run this command:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># named vector of locations</span>
<span class="va">locs</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">isolate_id</span>, <span class="va">facility</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">deframe</span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span></code></pre></div>
<pre><code>## PCMP_H1 PCMP_H2 PCMP_H3 PCMP_H4 PCMP_H5 PCMP_H7 
##     "A"     "B"     "C"     "D"     "E"     "G"</code></pre>
<p>The sample names are PCMP_H1, PCMP_H2, etc. And the facilities are A, B, etc.</p>
</div>
</div>
<div id="overview-of-questions" class="section level1">
<h1 class="hasAnchor">
<a href="#overview-of-questions" class="anchor"></a>Overview of questions</h1>
<p>In the following vignette, we go over the following questions and how <code>regentrans</code> can be used to investigate them:</p>
<ol start="0" style="list-style-type: decimal">
<li>How do you choose pairwise SNV distance thresholds?</li>
<li>Is transmission occurring within and/or between facilities?</li>
<li>What facilities is transmission occurring within/between?</li>
<li>Have transmission dynamics changed over time?</li>
<li>Is transmission occurring along paths of higher patient/person flow?</li>
<li>Are there any observable geographic trends in prevalence/transmission?</li>
</ol>
</div>
<div id="how-do-you-choose-pairwise-snv-distance-thresholds" class="section level1">
<h1 class="hasAnchor">
<a href="#how-do-you-choose-pairwise-snv-distance-thresholds" class="anchor"></a>0. How do you choose pairwise SNV distance thresholds?</h1>
<p>While some of the methods we present here are threshold-free, some require you to choose a pairwise SNV distance threshold. We highly recommend doing a sensitivity analysis with several thresholds to see how robust the results are to different thresholds.</p>
<p>To help you choose reasonable pairwise SNV distance thresholds for a sensitivity analysis, we describe two methods below.</p>
<div id="method-1-use-reference-genome-length-and-mutation-rate" class="section level2">
<h2 class="hasAnchor">
<a href="#method-1-use-reference-genome-length-and-mutation-rate" class="anchor"></a>Method 1: Use reference genome length and mutation rate</h2>
<p>One way to choose a pairwise SNV distance threshold is to use the estimated mutation rate of the organism you are studying. Here is an example for our dataset (<em>K. pneumoniae</em> ST258):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ref_genome_length</span> <span class="op">&lt;-</span> <span class="fl">5394056</span>
<span class="va">mutation_rate</span> <span class="op">&lt;-</span> <span class="fl">1.03e-6</span> <span class="co"># estimated from a K. pneumoniae ST258 time tree</span>
<span class="fu"><a href="https://rdrr.io/r/base/Round.html">floor</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">ref_genome_length</span><span class="op">*</span><span class="va">mutation_rate</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 11</code></pre>
<p>This value means that, over the course of a year, each ST258 strain is expected to gain about 5.5 mutations. This gives us a parwise SNV distance threshold of <span class="math inline">\(5.5*2 = 11\)</span>, as our isolates were collected over a year-long period.</p>
</div>
<div id="method-2-visualizing-intra-facility-pair-fraction-distribution-with-help-from-get_frac_intra" class="section level2">
<h2 class="hasAnchor">
<a href="#method-2-visualizing-intra-facility-pair-fraction-distribution-with-help-from-get_frac_intra" class="anchor"></a>Method 2: Visualizing intra-facility pair fraction distribution with help from <code>get_frac_intra()</code>
</h2>
<p>Sometimes it is difficult to calculate a mutation rate for a given organism, or the mutation rate might be variable even within a clonal lineage. Therefore, we also recommend using the data to identify potential pairwise SNV distance thresholds. Here, we use the assumption that intra-facility transmission is more common than inter-facility transmission to identify reasonable thresholds by looking for drops in the fraction of isolate pairs that are intra-facility pairs compared to inter-facility pairs for various SNV distances. Note that not all datasets will drop off, so this method may not work for some datasets.</p>
<p>We can use <code><a href="../reference/get_pair_types.html">get_pair_types()</a></code> to get the pair types for each isolate pair (intra- or inter-facility) and <code><a href="../reference/get_frac_intra.html">get_frac_intra()</a></code> to calculate the fraction of intra-facility pairs for different SNV distances:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get pair types for pairwise SNV distances (intra vs. inter)</span>
<span class="va">pair_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_pair_types.html">get_pair_types</a></span><span class="op">(</span>dists <span class="op">=</span> <span class="va">dists</span>, locs <span class="op">=</span> <span class="va">locs</span><span class="op">)</span>

<span class="co"># get fraction of intra-facility pairs for each SNV distance</span>
<span class="va">frac_intra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_frac_intra.html">get_frac_intra</a></span><span class="op">(</span>pair_types <span class="op">=</span> <span class="va">pair_types</span><span class="op">)</span></code></pre></div>
<p>The output of <code><a href="../reference/get_pair_types.html">get_pair_types()</a></code> includes information about each isolate pair including the pairwise distance, location of each isolate in the pair, and the type of pair (intra- or inter-facility):</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pair_types</span><span class="op">)</span></code></pre></div>
<pre><code>##   isolate1 isolate2 pairwise_dist loc1 loc2           pair_type
## 1  PCMP_H2  PCMP_H1            42    B    A Inter-facility pair
## 2  PCMP_H3  PCMP_H1            91    C    A Inter-facility pair
## 3  PCMP_H4  PCMP_H1            65    D    A Inter-facility pair
## 4  PCMP_H5  PCMP_H1           148    E    A Inter-facility pair
## 5  PCMP_H7  PCMP_H1            87    G    A Inter-facility pair
## 6  PCMP_H8  PCMP_H1           201    H    A Inter-facility pair</code></pre>
<p>Each row of the <code><a href="../reference/get_frac_intra.html">get_frac_intra()</a></code> output is a different pairwise distance and the columns include the number and fraction of intra- and inter-facility pairs for each pairwise distance:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">frac_intra</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 x 5
## # Groups:   pairwise_dist [6]
##   pairwise_dist n_intra n_inter frac_intra frac_inter
##           &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
## 1             0      18       6      0.75      0.25  
## 2             1      23       2      0.92      0.0800
## 3             2      18      12      0.6       0.4   
## 4             3      23       4      0.852     0.148 
## 5             4      32       7      0.821     0.179 
## 6             5      34      10      0.773     0.227</code></pre>
<p>Below we plot the fraction of intra-facility pairs for pairwise SNV distance up to 50. We have highlighted two drops in fraction of intra-facility pairs, which may be indicative of potential reasonable pairwise SNV distance thresholds. On of these (≤ 10) matches closely with Method 1. Based on this data, we use thresholds of 10 and 6 for our sensitivity analysis.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot fraction of intra-facility pairs for each SNV distance</span>
<span class="va">frac_intra</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>under <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">pairwise_dist</span> <span class="op">&lt;=</span> <span class="fl">10</span>, <span class="st">'≤ 10 SNVs'</span>,<span class="st">'Over threshold'</span><span class="op">)</span>, under <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">pairwise_dist</span> <span class="op">&lt;=</span> <span class="fl">6</span>, <span class="st">'≤ 6 SNVs'</span>,<span class="va">under</span><span class="op">)</span>,
         under <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">under</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'≤ 6 SNVs'</span>, <span class="st">'≤ 10 SNVs'</span>,<span class="st">'Over threshold'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pairwise_dist</span>, y <span class="op">=</span> <span class="va">frac_intra</span>, fill <span class="op">=</span> <span class="va">under</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_fill_grey</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Pairwise SNV distance"</span>, y <span class="op">=</span> <span class="st">"Fraction of intra-facility pairs"</span>, fill <span class="op">=</span> <span class="st">'Possible SNV\nthresholds'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">51</span><span class="op">)</span> </code></pre></div>
<pre><code>## Warning: Removed 235 rows containing missing values (position_stack).</code></pre>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_bar).</code></pre>
<p><img src="Introduction_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>Our dataset is extremely comprehensive and the isolates are all very closely related. However, if this is not the case in your dataset, you may want to bin the x axis into groups of pairwise SNV distances:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">frac_intra_bin</span> <span class="op">&lt;-</span> <span class="va">frac_intra</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">pairwise_dist</span> <span class="op">&lt;=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>bin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html">cut</a></span><span class="op">(</span><span class="va">pairwise_dist</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">pairwise_dist</span><span class="op">)</span> <span class="op">+</span> <span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, right <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">group_by</span><span class="op">(</span><span class="va">bin</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">summarise</span><span class="op">(</span>n_intra <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n_intra</span><span class="op">)</span>, n_inter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n_inter</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>frac_intra <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">n_intra</span> <span class="op">!=</span> <span class="fl">0</span>,
                             <span class="va">n_intra</span><span class="op">/</span><span class="op">(</span><span class="va">n_intra</span><span class="op">+</span><span class="va">n_inter</span><span class="op">)</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span>


<span class="va">frac_intra_bin</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">bin</span>, y <span class="op">=</span> <span class="va">frac_intra</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_fill_grey</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Pairwise SNV distance"</span>, y <span class="op">=</span> <span class="st">"Fraction of intra-facility pairs"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span> </code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>In this case, we might have chosen a pairwise SNV distance of 10, or tried plotting the data with other bins to get a more nuanced understanding of how the fraction changes, if we have enough data to do so.</p>
</div>
</div>
<div id="is-transmission-occurring-within-andor-between-locations" class="section level1">
<h1 class="hasAnchor">
<a href="#is-transmission-occurring-within-andor-between-locations" class="anchor"></a>1. Is transmission occurring within and/or between locations?</h1>
<p>To explore this question we can leverage our whole-genome sequencing data in the form of the phylogenetic tree using the <code><a href="../reference/get_clusters.html">get_clusters()</a></code> function and the SNV distance matrix using the <code><a href="../reference/get_pair_types.html">get_pair_types()</a></code>.</p>
<div id="threshold-free-phylogenetic-approach" class="section level2">
<h2 class="hasAnchor">
<a href="#threshold-free-phylogenetic-approach" class="anchor"></a>Threshold-free phylogenetic approach</h2>
<div id="visualize-the-phylogenetic-tree-annotated-by-location" class="section level3">
<h3 class="hasAnchor">
<a href="#visualize-the-phylogenetic-tree-annotated-by-location" class="anchor"></a>Visualize the phylogenetic tree annotated by location</h3>
<p>It is useful to first visualize the extent of clustering on the phylogeny. To do this, we can plot the tree using ggtree:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># pick prettier colors than the default</span>
<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'#e6194b'</span>, <span class="st">'#3cb44b'</span>, <span class="st">'#ffe119'</span>, <span class="st">'#4363d8'</span>, <span class="st">'#f58231'</span>, <span class="st">'#911eb4'</span>, <span class="st">'#46f0f0'</span>, <span class="st">'#f032e6'</span>, <span class="st">'#bcf60c'</span>, <span class="st">'#fabebe'</span>, <span class="st">'#008080'</span>, <span class="st">'#e6beff'</span>, <span class="st">'#9a6324'</span>, <span class="st">'#fffac8'</span>, <span class="st">'#800000'</span>, <span class="st">'#aaffc3'</span>, <span class="st">'#808000'</span>, <span class="st">'#ffd8b1'</span>, <span class="st">'#000075'</span>, <span class="st">'#808080'</span>, <span class="st">'#000000'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span>

<span class="co"># create vector for plotting location as colors</span>
<span class="va">locs_tip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">locs</span><span class="op">[</span><span class="va">tr</span><span class="op">$</span><span class="va">tip.label</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/pkg/ape/man/summary.phylo.html">Nnode</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># plot tree</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">locs_tip</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>col <span class="op">=</span> <span class="st">'Facility'</span><span class="op">)</span> </code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>As we can see here, isolates from some facilities tend to cluster on the phylogeny (e.g. facilities D and E), which suggests that intra-facility transmission may be occurrring.</p>
</div>
<div id="quantify-clusters-using-get_clusters" class="section level3">
<h3 class="hasAnchor">
<a href="#quantify-clusters-using-get_clusters" class="anchor"></a>Quantify clusters using <code>get_clusters()</code>
</h3>
<p>It is also useful to quantify the size of phylogenetic clusters from the same facility. Larger clusters indicate more within-facility transmission while smaller clusters or singletons indicate more importation. When working with the <code><a href="../reference/get_clusters.html">get_clusters()</a></code> function, you may perform analyses with various cluster pureness levels. The default pureness value is 1. Lower values will allow clusters to be identified that have some “contamination” with other facilities.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get clusters (note this takes some time to run, but not too long for the test dataset!)</span>
<span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_clusters.html">get_clusters</a></span><span class="op">(</span><span class="va">tr</span>,<span class="va">locs</span>, pureness <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>The output of <code><a href="../reference/get_clusters.html">get_clusters()</a></code> is a list including a data frame of information about the pure subtrees identifed and a list of the subtrees themselves:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dissect output</span>
<span class="va">pure_subtree_info</span> <span class="op">&lt;-</span> <span class="va">clusters</span><span class="op">$</span><span class="va">pure_subtree_info</span>
<span class="va">subtrees</span> <span class="op">&lt;-</span> <span class="va">clusters</span><span class="op">$</span><span class="va">subtrees</span>
<span class="va">pure_subtree_info</span></code></pre></div>
<pre><code>## # A tibble: 224 x 4
##    loc   subtr_size index isolate_id
##    &lt;chr&gt;      &lt;dbl&gt; &lt;int&gt; &lt;chr&gt;     
##  1 D              1    NA PCMP_H100 
##  2 N              1    NA PCMP_H98  
##  3 D              1    NA PCMP_H141 
##  4 D              3    10 &lt;NA&gt;      
##  5 D              1    NA PCMP_H142 
##  6 E              6    17 &lt;NA&gt;      
##  7 H              1    NA PCMP_H391 
##  8 D              1    NA PCMP_H93  
##  9 H              1    NA PCMP_H138 
## 10 T              1    NA PCMP_H281 
## # … with 214 more rows</code></pre>
<p>The data frame includes the location, the size of the subtree (i.e. cluster), the index of the subtree in the list of subtrees, and the isolate ID if it is a singleton.</p>
<p>We can plot the size of these clusters to visualize the output:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">pure_subtree_info</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">loc</span>, y <span class="op">=</span> <span class="va">subtr_size</span>, color <span class="op">=</span> <span class="va">loc</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_jitter</span><span class="op">(</span>position <span class="op">=</span> <span class="fu">position_jitter</span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span>, height <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Number of isolates in \nphylogenetic cluster"</span>, x <span class="op">=</span> <span class="st">""</span>, color <span class="op">=</span> <span class="st">'Facility'</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">12.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">coord_flip</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>Here, we confirm what we saw visually - that some facilities have large pure clusters of isolates, suggesting intra-facility transmission.</p>
</div>
<div id="assign-each-isolate-to-a-cluster" class="section level3">
<h3 class="hasAnchor">
<a href="#assign-each-isolate-to-a-cluster" class="anchor"></a>Assign each isolate to a cluster</h3>
<p>Sometimes it’s useful to know which cluster each isolate belongs to. To do this, you can use the following code:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get ids of isolates in each cluster</span>
<span class="va">isolates_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">pure_subtree_info</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span>
  <span class="va">i</span> <span class="op">&lt;-</span> <span class="va">pure_subtree_info</span><span class="op">$</span><span class="va">index</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>
  <span class="va">name</span> <span class="op">&lt;-</span> <span class="va">pure_subtree_info</span><span class="op">$</span><span class="va">isolate_id</span><span class="op">[</span><span class="va">x</span><span class="op">]</span>
  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">name</span> <span class="op">&lt;-</span> <span class="va">subtrees</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">tip.label</span>
  <span class="op">}</span>
  <span class="va">name</span>
<span class="op">}</span><span class="op">)</span>

<span class="co"># get unique cluster names</span>
<span class="va">facil_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span>
  <span class="va">clusts</span> <span class="op">&lt;-</span> <span class="va">isolates_clusters</span><span class="op">[</span><span class="va">pure_subtree_info</span><span class="op">$</span><span class="va">loc</span> <span class="op">==</span> <span class="va">x</span><span class="op">]</span>
  <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">clusts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">clusts</span><span class="op">)</span><span class="op">)</span>
  <span class="va">cluster_nums</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">clusts</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">clusts</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
  <span class="op">}</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cluster_nums</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">clusts</span><span class="op">)</span>
  <span class="va">cluster_nums</span>
<span class="op">}</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/unname.html">unname</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">facil_clusters</span><span class="op">)</span></code></pre></div>
<pre><code>## PCMP_H144  PCMP_H91   PCMP_H1 PCMP_H326  PCMP_H49 PCMP_H156 
##      "A1"      "A1"      "A1"      "A2"      "A3"      "A4"</code></pre>
<p><code>facil_clusters</code> is a named list where the names are the isolate id and the elements are the cluster that it belongs to, named by facility followed by cluster number for that facility.</p>
<p>With this many groups, it’s kind of crazy to visualize on the phylogeny, but it could be useful if you have fewer:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clust_tips</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">facil_clusters</span><span class="op">[</span><span class="va">tr</span><span class="op">$</span><span class="va">tip.label</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/pkg/ape/man/summary.phylo.html">Nnode</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tippoint.html">geom_tippoint</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">clust_tips</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span> </code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
</div>
<div id="pairwise-snv-distance-approach" class="section level2">
<h2 class="hasAnchor">
<a href="#pairwise-snv-distance-approach" class="anchor"></a>Pairwise SNV distance approach</h2>
<p>Even if isolates from the same facility cluster on the phylogeny, they may still be distantly related to each other. To see how closely related isolate pairs are, we can leverage our whole-genome sequencing data in the form of the SNV distance matrix and the <code><a href="../reference/get_pair_types.html">get_pair_types()</a></code> function. We already identified intra- and inter-facility pairs above using <code><a href="../reference/get_pair_types.html">get_pair_types()</a></code>, so we won’t re-compute them here.</p>
<p>Below we plot the intra- and inter-facility pairwise SNV distances as a histogram. The grey line is a pairwise SNV distance of 11. We can see that there are closely related intra- and inter-facility isolate pairs, with an enrichment in intra-facility pairs.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snv_hist</span> <span class="op">&lt;-</span> <span class="va">pair_types</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pairwise_dist</span>, fill <span class="op">=</span> <span class="va">pair_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_histogram</span><span class="op">(</span>position <span class="op">=</span> <span class="st">'identity'</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>, bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Pairwise SNV distance"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, fill <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">11</span>, col <span class="op">=</span> <span class="st">'darkgrey'</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.justification <span class="op">=</span> <span class="st">"top"</span>, 
        axis.line <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,
        panel.border <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">snv_hist</span></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>We can also take a closer look at the closely related pairs by zooming into the left side of the histogram, further highlighting the enrichment in intra-facility pairs at small pairwise SNV distances:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snv_hist</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">50</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Removed 49782 rows containing non-finite values (stat_bin).</code></pre>
<pre><code>## Warning: Removed 4 rows containing missing values (geom_bar).</code></pre>
<p><img src="Introduction_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<div id="is-transmission-occurring-between-different-sample-sources" class="section level4">
<h4 class="hasAnchor">
<a href="#is-transmission-occurring-between-different-sample-sources" class="anchor"></a>Is transmission occurring between different sample sources?</h4>
<p>If your dataset includes environmental samples and, you could also explore whether transmission is occurring between different sample sources by considering each environmental source a location. You could then determine what fraction of your isolate pairs are patient-patient pairs vs. patient-environment pairs vs. environment-environment pairs with the code below.</p>
<p>Note that this is a <strong>toy example</strong> as this sample data only includes patient samples.</p>
<p>If your samples are labeled differently (i.e. patient samples are labeled with PT_… and environmental samples are labeled with EN_…) you may want to use a function like <code>grepl</code> to differentiate between sample types.</p>
<p>Note that this is one of many ways in which the functions provided in <code>regentrans</code> can be used to ask questions about transmission that we do not directly discuss here. Other examples include looking at transmission between zip codes or wards/rooms in a facility.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create toy data</span>
<span class="co"># each patient and each environment is a different location</span>
<span class="co"># get patients </span>
<span class="va">pts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span>
<span class="co"># get environments</span>
<span class="va">envs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span><span class="op">[</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">locs</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="va">pair_types</span> <span class="op">%&gt;%</span> <span class="fu">mutate</span><span class="op">(</span>src1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">loc1</span> <span class="op">%in%</span> <span class="va">pts</span>, <span class="st">'pt'</span>,<span class="st">'env'</span><span class="op">)</span>,
                     src2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">loc2</span> <span class="op">%in%</span> <span class="va">pts</span>, <span class="st">'pt'</span>,<span class="st">'env'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pairwise_dist</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">src1</span>, <span class="va">src2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_histogram</span><span class="op">(</span>position <span class="op">=</span> <span class="st">'identity'</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>, bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Pairwise SNV distance"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, fill <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">11</span>, col <span class="op">=</span> <span class="st">'darkgrey'</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
</div>
</div>
<div id="what-locations-is-transmission-occurring-withinbetween" class="section level1">
<h1 class="hasAnchor">
<a href="#what-locations-is-transmission-occurring-withinbetween" class="anchor"></a>2. What locations is transmission occurring within/between?</h1>
<p>We can use threshold-free or SNV threshold-based approaches to investigate what locations transmission is occurring within or between.</p>
<div id="gene-flow-fsp" class="section level3">
<h3 class="hasAnchor">
<a href="#gene-flow-fsp" class="anchor"></a>Gene flow (Fsp)</h3>
<p>To determine the extent of gene flow between two facilities, we can calculate Fsp (<a href="mgen.microbiologyresearch.org/pubmed/content/journal/mgen/10.1099/mgen.0.000113">Donker et al., 2017</a>) as a measure of similarity between populations at different facilities. A value of 0 indicates that the populations are identical, while value of 1 indicates that the populations are completely different. The fsp function outputs a symmetric matrix of fsp values for each facility pair. The row and column names are the facilities and the values are fsp:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fsp</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></code></pre></div>
<pre><code>##           A         C         D         E
## A 0.0000000 0.7679634 0.7469595 0.8197587
## C 0.7679634 0.0000000 0.3525517 0.3523582
## D 0.7469595 0.3525517 0.0000000 0.2267869
## E 0.8197587 0.3523582 0.2267869 0.0000000</code></pre>
<p>We can plot the fsp values as a heatmap to visualze the extent of similarity between strains from different facility pairs:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this takes a while to run, so we've pre-computed it and included it in the package</span>
<span class="co"># fsp &lt;- get_facility_fsp(aln, locs)</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">fsp</span><span class="op">)</span></code></pre></div>
<p>Here we can see that some facilities have more similar CRKP populations than others, suggesting that there is more transmission between those facilities.</p>
<p>One way to determine whether fsp values are significant is to use permutation (<a href="mgen.microbiologyresearch.org/pubmed/content/journal/mgen/10.1099/mgen.0.000113">Donker et al., 2017</a>).</p>
</div>
<div id="snv-distances" class="section level3">
<h3 class="hasAnchor">
<a href="#snv-distances" class="anchor"></a>SNV distances</h3>
<p>We can also look at the number of closely related pairs of isolates at different facility pairs to get a sense of the extent of potential recent transmission and identify facilities and facility pairs with more putative spread. For this, we must choose a SNV threshold. Therefore, we recommend performing a sensitivity analysis by choosing several different SNV thresholds and seeing how robust the results are to these changes.</p>
<p>To do this, we can summarize the pairwise SNV distance information using the <code><a href="../reference/summarize_pairs.html">summarize_pairs()</a></code> function:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get summary of pairwise SNV distance data</span>
<span class="va">snv_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_pairs.html">summarize_pairs</a></span><span class="op">(</span><span class="va">pair_types</span>,summary_fns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'min'</span><span class="op">)</span>, threshs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span> 
<span class="va">snv_summary</span></code></pre></div>
<pre><code>## # A tibble: 227 x 5
## # Groups:   loc1, loc2 [227]
##    loc1  loc2  dist_min leq_6 leq_10
##    &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;int&gt;  &lt;int&gt;
##  1 A     A            1     3      3
##  2 A     B           19     0      0
##  3 A     C           63     0      0
##  4 A     D           36     0      0
##  5 A     E           65     0      0
##  6 A     F           70     0      0
##  7 A     G           69     0      0
##  8 A     H           71     0      0
##  9 A     I           73     0      0
## 10 A     J           42     0      0
## # … with 217 more rows</code></pre>
<p>The output is a summary of the isolate pairs for different facility pairs. Here, we also filtered out intra-facility pairs, as we are most interested in inter-facility transmission as compared to patient transfer. The function allows you to input a vector of whatever summary functions you’d like (default is minimum pairwise SNV distance), as well as pairwise SNV distance thresholds to compute the number of closely related pairs for each facility pair under that threshold.</p>
<p>Now, we can plot that information to identify facilities with closely related intra- and inter-facility pairs:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get facility summary</span>
<span class="va">facil_summary</span> <span class="op">&lt;-</span> <span class="va">snv_summary</span> <span class="op">%&gt;%</span> 
    <span class="fu">mutate</span><span class="op">(</span>pair_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">loc1</span> <span class="op">==</span> <span class="va">loc2</span>, <span class="st">'Intra-facility pair'</span>, <span class="st">'Inter-facility pair'</span><span class="op">)</span>, 
                            levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Intra-facility pair'</span>, <span class="st">'Inter-facility pair'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">group_by</span><span class="op">(</span><span class="va">loc1</span>, <span class="va">pair_type</span>, <span class="va">leq_6</span>, <span class="va">leq_10</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">summarize</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">leq_6</span>, <span class="va">leq_10</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">rename</span><span class="op">(</span>thresh <span class="op">=</span> <span class="va">name</span>, n_pairs <span class="op">=</span> <span class="va">value</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>thresh <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">'leq_'</span>,<span class="st">'≤ '</span>, <span class="va">thresh</span><span class="op">)</span>, <span class="st">'SNVs'</span><span class="op">)</span><span class="op">)</span> 

<span class="co"># get locations to keep (at least 1 closely related pair)</span>
<span class="va">keep_locs</span> <span class="op">&lt;-</span> <span class="va">facil_summary</span> <span class="op">%&gt;%</span> <span class="fu">group_by</span><span class="op">(</span><span class="va">loc1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">summarize</span><span class="op">(</span>sum_all <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n_pairs</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">sum_all</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">loc1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">deframe</span><span class="op">(</span><span class="op">)</span>

<span class="co"># plot</span>
<span class="va">facil_summary</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">loc1</span> <span class="op">%in%</span> <span class="va">keep_locs</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/complete.html">complete</a></span><span class="op">(</span><span class="va">loc1</span>, <span class="va">pair_type</span>, <span class="va">thresh</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html">reorder</a></span><span class="op">(</span><span class="va">loc1</span>, <span class="op">-</span><span class="va">n_pairs</span>, <span class="va">sum</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">n_pairs</span>, fill <span class="op">=</span> <span class="va">pair_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_col</span><span class="op">(</span>position <span class="op">=</span> <span class="st">'dodge'</span><span class="op">)</span> <span class="op">+</span>
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cadetblue"</span>, <span class="st">"salmon"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
         <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Number of closely related \nisolate pairs"</span>, x <span class="op">=</span> <span class="st">"Facility"</span>, fill <span class="op">=</span> <span class="st">''</span>, color <span class="op">=</span> <span class="st">''</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_grid</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">thresh</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'≤ 6 SNVs'</span>,<span class="st">'≤ 10 SNVs'</span><span class="op">)</span><span class="op">)</span><span class="op">~</span><span class="va">.</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Removed 2 rows containing missing values (geom_col).</code></pre>
<p><img src="Introduction_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
<p>Here we see that there is a large number of closely related isolates at certain facilities (e.g. D) and very few at others. We also see that, while most facilities have more closely related intra-facility pairs than inter-facility pairs, this observation does not hold for facilities C and F, which has more closely related intra-facility pairs than inter-facility pairs.</p>
<p>Furthermore, the results vary somewhat between the two different SNV thresholds, but are relatively consistent.</p>
</div>
</div>
<div id="have-transmission-dynamics-changed-over-time-or-across-species" class="section level1">
<h1 class="hasAnchor">
<a href="#have-transmission-dynamics-changed-over-time-or-across-species" class="anchor"></a>3. Have transmission dynamics changed over time or across species?</h1>
<p>To investigate whether transmission dynamics have changed over time or differ across groups (such as sequence type or species) we can “facet” on those variables. Here is an example of faceting on date:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get dates </span>
<span class="va">dates</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">isolate_id</span>, <span class="va">collection_date</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">deframe</span><span class="op">(</span><span class="op">)</span>

<span class="va">time1</span> <span class="op">&lt;-</span> <span class="va">pair_types</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>date1 <span class="op">=</span> <span class="va">dates</span><span class="op">[</span><span class="va">isolate1</span><span class="op">]</span>, date2 <span class="op">=</span> <span class="va">dates</span><span class="op">[</span><span class="va">isolate2</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">date1</span> <span class="op">==</span> <span class="va">date2</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pairwise_dist</span>, fill <span class="op">=</span> <span class="va">pair_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_histogram</span><span class="op">(</span>position <span class="op">=</span> <span class="st">'identity'</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>, bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Pairwise SNV distance"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, fill <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span><span class="va">date1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.justification <span class="op">=</span> <span class="st">"top"</span>, 
        axis.line <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,
        panel.border <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="va">time2</span> <span class="op">&lt;-</span> <span class="va">pair_types</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>date1 <span class="op">=</span> <span class="va">dates</span><span class="op">[</span><span class="va">isolate1</span><span class="op">]</span>, date2 <span class="op">=</span> <span class="va">dates</span><span class="op">[</span><span class="va">isolate2</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">date1</span> <span class="op">==</span> <span class="va">date2</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">pairwise_dist</span>, fill <span class="op">=</span> <span class="va">pair_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_histogram</span><span class="op">(</span>position <span class="op">=</span> <span class="fu">position_fill</span><span class="op">(</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>, bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span>
  <span class="co"># xlim(0,20) +</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Pairwise SNV distance"</span>, y <span class="op">=</span> <span class="st">"Fraction"</span>, fill <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span><span class="va">date1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.justification <span class="op">=</span> <span class="st">"top"</span>, 
        axis.line <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,
        panel.border <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">time1</span>, <span class="va">time2</span>, nrow <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">'AUTO'</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Removed 8 rows containing missing values (geom_bar).</code></pre>
<p><img src="Introduction_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<p>The first row of plots show histograms of the raw counts while the second row of plots shows the fraction of isolates at that pairwise SNV distance. We don’t observe many differences over time, except that there are more isolate pairs in 2015 compared to 2014. This is not unexpected given that the dataset we are using is from an endemic setting. If the dataset is from an outbreak setting, you may expect changes in closely related intra- and inter-facility pairs over time. For example, an increase in intra-facility transmission followed by an increase in inter-facility transmission as it disseminates across the healthcare network.</p>
</div>
<div id="is-transmission-occurring-along-paths-of-higher-patient-flow" class="section level1">
<h1 class="hasAnchor">
<a href="#is-transmission-occurring-along-paths-of-higher-patient-flow" class="anchor"></a>4. Is transmission occurring along paths of higher patient flow?</h1>
<p>To explore this question, we can leverage our whole-genome sequencing data in the form of the SNV distance matrix and/or Fsp matrix along with the patient transfer network using the <code><a href="../reference/get_patient_flow.html">get_patient_flow()</a></code>, <code><a href="../reference/summarize_pairs.html">summarize_pairs()</a></code>, and <code><a href="../reference/merge_inter_summaries.html">merge_inter_summaries()</a></code> functions.</p>
<p><em>Note:</em> If you do not have a patient transfer network, you may be able to use geographic distances between facilities as a proxy for patient transfer because they are often correlated.</p>
<p><strong>Remember that choosing pairwise SNV distance threshold(s) based on your dataset is very important, avoid using the default values for your own dataset.</strong></p>
<p>We already summarized the genomic data above using <code><a href="../reference/summarize_pairs.html">summarize_pairs()</a></code>.</p>
<p>We can also get the direct and indirect patient flow between facilities (see <a href="https://aac.asm.org/content/63/11/e01622-19">Han et al. 2019</a> for details on how indirect flow is calculated.) using <code>get_patinet_flow()</code>. The input to this is a data frame with three columns:</p>
<ol style="list-style-type: decimal">
<li>
<code>source_facil</code>: the source facility</li>
<li>
<code>dest_facil</code>: the destination facility</li>
<li>
<code>n_transfers</code>: the number of transfers from source facility to destination facility over a certain amount of time (e.g. 1 month, 1 year, etc.)</li>
</ol>
<p>Here is a toy example that we generated for the vignette:</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pt_trans_df</span><span class="op">)</span></code></pre></div>
<pre><code>##   source_facil dest_facil n_transfers
## 1            H          D          17
## 2            G          D          14
## 3            C          D         474
## 4            T          D         182
## 5            F          D          60
## 6            I          D          63</code></pre>
<p>Here is the output of <code>get_patinet_flow()</code>:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pt_trans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_patient_flow.html">get_patient_flow</a></span><span class="op">(</span>pt_trans_df <span class="op">=</span> <span class="va">pt_trans_df</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pt_trans</span><span class="op">)</span></code></pre></div>
<pre><code>##   loc1 loc2 n_transfers_f12 pt_trans_metric_f12 n_transfers_f21
## 1    D    H             260          0.04512957              17
## 2    D    G               5          0.03212340              14
## 3    C    D             474          0.06218840            2660
## 4    D    T               8          0.02663025             182
## 5    D    F              46          0.06683914              60
## 6    D    I              12          0.16451422              63
##   pt_trans_metric_f21 sum_transfers sum_pt_trans_metric
## 1         0.019319187           277          0.06444875
## 2         0.004199755            19          0.03632315
## 3         0.391118953          3134          0.45330736
## 4         0.032739701           190          0.05936995
## 5         0.028057366           106          0.09489650
## 6         0.023562553            75          0.18807677</code></pre>
<p>The <code><a href="../reference/get_patient_flow.html">get_patient_flow()</a></code> function outputs several columns:</p>
<ul>
<li>The facility pair (<code>loc1</code> and <code>loc2</code>)</li>
<li>The number of transfers from facility 1 to 2 (<code>n_transfers_f12</code>), and facility 2 to 1 (<code>n_transfers_f21</code>)</li>
<li>The indirect flow metrics from facility 1 to 2 (<code>pt_trans_metric_f12</code>), and facility 2 to 1 (<code>pt_trans_metric_f21</code>)</li>
<li>Undirected transfer and flow metrics that are the sum of the two direct metrics (<code>sum_transfers</code> and <code>sum_pt_trans_metric</code>). This is useful for comparing to the undirected pairwise SNV distance summary and Fsp.</li>
</ul>
<p>Next, we can merge the patient flow, pairwise SNV distance, and fsp summaries (note that there are a lot of NA values here because we do not have patient flow information for all facilities):</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># have to convert fsp matrix to long form first</span>
<span class="va">fsp_long</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_long_form.html">make_long_form</a></span><span class="op">(</span><span class="va">fsp</span><span class="op">)</span>

<span class="co"># only keep inter-facility info in snv_summary</span>
<span class="va">inter_snv_summary</span> <span class="op">&lt;-</span> <span class="va">snv_summary</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">loc1</span> <span class="op">!=</span> <span class="va">loc2</span><span class="op">)</span>


<span class="va">pair_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_inter_summaries.html">merge_inter_summaries</a></span><span class="op">(</span><span class="va">pt_trans</span>, <span class="va">inter_snv_summary</span>, <span class="va">fsp_long</span><span class="op">)</span> 
<span class="va">pair_info</span></code></pre></div>
<pre><code>## # A tibble: 265 x 12
## # Groups:   loc1, loc2 [210]
##    loc1  loc2  dist_min leq_6 leq_10    fsp n_transfers_f12 pt_trans_metric_f12
##    &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt; &lt;int&gt;  &lt;int&gt;  &lt;dbl&gt;           &lt;dbl&gt;               &lt;dbl&gt;
##  1 A     B           19     0      0 NA                  NA                  NA
##  2 A     C           63     0      0  0.768              NA                  NA
##  3 A     D           36     0      0  0.747              NA                  NA
##  4 A     E           65     0      0  0.820              NA                  NA
##  5 A     F           70     0      0  0.739              NA                  NA
##  6 A     G           69     0      0  0.732              NA                  NA
##  7 A     H           71     0      0  0.844              NA                  NA
##  8 A     I           73     0      0  0.868              NA                  NA
##  9 A     J           42     0      0  0.974              NA                  NA
## 10 A     K          102     0      0  0.948              NA                  NA
## # … with 255 more rows, and 4 more variables: n_transfers_f21 &lt;dbl&gt;,
## #   pt_trans_metric_f21 &lt;dbl&gt;, sum_transfers &lt;dbl&gt;, sum_pt_trans_metric &lt;dbl&gt;</code></pre>
<p>Now we’re ready to plot patient flow vs. various genomic metrics of similarity between facilities:</p>
<p>We see a negative correlation between patient flow and Fsp, suggesting that facilities connected by more patient transfer are more likely to have similar populations, indicative of more transmission (Remember that fsp values closer to 0 indicate more closely related populations):</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pair_info</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">sum_pt_trans_metric</span>,y<span class="op">=</span><span class="va">fsp</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_smooth</span><span class="op">(</span>method<span class="op">=</span><span class="st">'lm'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">'Patient flow'</span>, y <span class="op">=</span> <span class="st">'Fsp'</span><span class="op">)</span> </code></pre></div>
<pre><code>## `geom_smooth()` using formula 'y ~ x'</code></pre>
<pre><code>## Warning: Removed 155 rows containing non-finite values (stat_smooth).</code></pre>
<pre><code>## Warning: Removed 155 rows containing missing values (geom_point).</code></pre>
<p><img src="Introduction_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<p>These observations are corroborated by the positive relationship between patient transfer and number of closely related isolate pairs for a given facility pair:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pair_info</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">sum_pt_trans_metric</span>,y<span class="op">=</span><span class="va">leq_10</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_smooth</span><span class="op">(</span>method<span class="op">=</span><span class="st">'lm'</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_x_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">'Patient flow'</span>, y <span class="op">=</span> <span class="st">'# closely related\npairs (≤ 10 SNVs)'</span><span class="op">)</span> </code></pre></div>
<pre><code>## `geom_smooth()` using formula 'y ~ x'</code></pre>
<pre><code>## Warning: Removed 155 rows containing non-finite values (stat_smooth).</code></pre>
<pre><code>## Warning: Removed 155 rows containing missing values (geom_point).</code></pre>
<p><img src="Introduction_files/figure-html/unnamed-chunk-39-1.png" width="700"></p>
</div>
<div id="are-there-any-observable-geographic-trends-in-prevalencetransmission" class="section level1">
<h1 class="hasAnchor">
<a href="#are-there-any-observable-geographic-trends-in-prevalencetransmission" class="anchor"></a>Are there any observable geographic trends in prevalence/transmission?</h1>
<p>To evaluate whether there are any observable geographic trends in prevalence or transmission, we will need geographic information (latitude and longitude) for each facility. Here we’ve de-identified the facility geographies as an example. Note that we don’t have geographic information for all facility pairs.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">facil_coord</span></code></pre></div>
<pre><code>## # A tibble: 11 x 3
##    facil     long      lat
##    &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;
##  1 C     -930605. -740730.
##  2 D     -930605. -740730.
##  3 E     -930605. -740730.
##  4 F     -930605. -740730.
##  5 G     -930605. -740730.
##  6 H     -930605. -740730.
##  7 I     -930605. -740730.
##  8 M     -930605. -740730.
##  9 O     -930605. -740729.
## 10 P     -930605. -740730.
## 11 T     -930605. -740730.</code></pre>
<p>This data frame contains the facility ID, longitude, latitude.</p>
<p>We can also add information about the number of isolates from each facility to include as point size on our map:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get counts for each location</span>
<span class="va">loc_n</span> <span class="op">&lt;-</span> <span class="va">locs</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">`colnames&lt;-`</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"facil"</span>, <span class="st">"n_isolates"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># merge longitude/latitude with locations</span>
<span class="va">facil_geo</span> <span class="op">&lt;-</span> <span class="va">facil_coord</span> <span class="op">%&gt;%</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">loc_n</span>, by <span class="op">=</span> <span class="st">'facil'</span><span class="op">)</span>
<span class="va">facil_geo</span></code></pre></div>
<pre><code>## # A tibble: 11 x 4
##    facil     long      lat n_isolates
##    &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;      &lt;int&gt;
##  1 C     -930605. -740730.         30
##  2 D     -930605. -740730.         38
##  3 E     -930605. -740730.         51
##  4 F     -930605. -740730.         19
##  5 G     -930605. -740730.         37
##  6 H     -930605. -740730.         29
##  7 I     -930605. -740730.         11
##  8 M     -930605. -740730.         33
##  9 O     -930605. -740729.         13
## 10 P     -930605. -740730.         24
## 11 T     -930605. -740730.         15</code></pre>
<p>We can use this to plot the facilities on a map and connect them by closely related pairs (here we use the under 7 threshold for visualization purposes). Point size represents number of isolates collected from a facility and transparency represents the number of closely related isolate pairs.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># merge facil_info and facil_geo</span>
<span class="va">snv_geo</span> <span class="op">&lt;-</span> <span class="va">pair_info</span> <span class="op">%&gt;%</span> 
  <span class="fu">left_join</span><span class="op">(</span><span class="va">facil_geo</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"loc1"</span> <span class="op">=</span> <span class="st">"facil"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">rename</span><span class="op">(</span>facil_1_lat <span class="op">=</span> <span class="va">lat</span>, facil_1_long <span class="op">=</span> <span class="va">long</span>, facil_1_n <span class="op">=</span> <span class="va">n_isolates</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">left_join</span><span class="op">(</span><span class="va">facil_geo</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"loc2"</span> <span class="op">=</span> <span class="st">"facil"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">rename</span><span class="op">(</span>facil_2_lat <span class="op">=</span> <span class="va">lat</span>, facil_2_long <span class="op">=</span> <span class="va">long</span>, facil_2_n <span class="op">=</span> <span class="va">n_isolates</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">facil_1_lat</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">facil_2_lat</span><span class="op">)</span><span class="op">)</span>

<span class="co"># filter to not plot zeros</span>
<span class="va">snv_geo_thresh</span> <span class="op">&lt;-</span> <span class="va">snv_geo</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">leq_6</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> 

<span class="va">facil_geo</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">long</span>, y<span class="op">=</span><span class="va">lat</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_curve</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">facil_1_long</span>, y <span class="op">=</span> <span class="va">facil_1_lat</span>, xend <span class="op">=</span> <span class="va">facil_2_long</span>, yend <span class="op">=</span> <span class="va">facil_2_lat</span>,
                 alpha <span class="op">=</span> <span class="va">leq_6</span><span class="op">)</span>,
             data <span class="op">=</span> <span class="va">snv_geo_thresh</span>, curvature <span class="op">=</span> <span class="fl">0.33</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size<span class="op">=</span><span class="va">n_isolates</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">1</span>, color <span class="op">=</span> <span class="st">"lightblue"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_map.html">theme_map</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">labs</span><span class="op">(</span>size <span class="op">=</span> <span class="st">'Number of samples'</span>, alpha <span class="op">=</span> <span class="st">'Number of closely related\npairs (≤ 6 SNVs)'</span><span class="op">)</span></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-42-1.png" width="700"></p>
<p>We can also look at the relationship between physical distance between facilities and number of closely related pairs. First, we have to get the physical distance between facilities:</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snv_geo</span> <span class="op">&lt;-</span> <span class="va">snv_geo</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>geo_dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="op">(</span><span class="va">facil_1_long</span><span class="op">-</span><span class="va">facil_2_long</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">+</span><span class="op">(</span><span class="va">facil_1_lat</span><span class="op">-</span><span class="va">facil_2_lat</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Then we can plot the relationship between physical distance and summary metrics. Here is an example using the number of pairs under a pairwise SNV distance of 11 between facility pairs:</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snv_geo</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">geo_dist</span>, y <span class="op">=</span> <span class="va">leq_10</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">'lm'</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">'Physical distance between facilities'</span>, y <span class="op">=</span> <span class="st">'# closely related pairs\n(≤ 10 SNVs)'</span><span class="op">)</span></code></pre></div>
<pre><code>## `geom_smooth()` using formula 'y ~ x'</code></pre>
<p><img src="Introduction_files/figure-html/unnamed-chunk-44-1.png" width="700"></p>
</div>
<div id="a-note-on-additional-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#a-note-on-additional-metadata" class="anchor"></a>A note on additional metadata</h1>
<p>Sometimes it’s useful to make plots faceted or subsetted based on various information from metadata such as collection date, sequence type, etc. To do this, you can merge the <code>pair_types</code> object with the <code>metadata</code> object as follows (modifying the code as fit for your dataset):</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># merge pair_types and metadata, and find time difference</span>
<span class="va">snv_and_metadat</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">pair_types</span>, <span class="va">metadata</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'isolate1'</span> <span class="op">=</span> <span class="st">'isolate_id'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">rename</span><span class="op">(</span>patient_id_1 <span class="op">=</span> <span class="va">patient_id</span>, collection_date_1 <span class="op">=</span> <span class="va">collection_date</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">left_join</span><span class="op">(</span><span class="va">metadata</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'isolate2'</span> <span class="op">=</span> <span class="st">'isolate_id'</span><span class="op">)</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">snv_and_metadat</span><span class="op">)</span></code></pre></div>
<pre><code>##   isolate1 isolate2 pairwise_dist loc1 loc2           pair_type patient_id_1
## 1  PCMP_H2  PCMP_H1            42    B    A Inter-facility pair           p2
## 2  PCMP_H3  PCMP_H1            91    C    A Inter-facility pair           p3
## 3  PCMP_H4  PCMP_H1            65    D    A Inter-facility pair           p4
## 4  PCMP_H5  PCMP_H1           148    E    A Inter-facility pair           p5
## 5  PCMP_H7  PCMP_H1            87    G    A Inter-facility pair           p7
## 6  PCMP_H8  PCMP_H1           201    H    A Inter-facility pair           p8
##   facility.x collection_date_1 patient_id facility.y collection_date
## 1          B              2014         p1          A            2014
## 2          C              2014         p1          A            2014
## 3          D              2014         p1          A            2014
## 4          E              2014         p1          A            2014
## 5          G              2014         p1          A            2014
## 6          H              2014         p1          A            2014</code></pre>
</div>
<div id="making-your-code-run-faster" class="section level1">
<h1 class="hasAnchor">
<a href="#making-your-code-run-faster" class="anchor"></a>Making your code run faster</h1>
<p>For some functions, we have used <code>future.apply()</code> to enable you to make it run faster. Currently, this is only an option for <code><a href="../reference/get_clusters.html">get_clusters()</a></code>, but we plan to implement it for <code><a href="../reference/get_facility_fsp.html">get_facility_fsp()</a></code> as well, and possibly others. To decrease runtime in this way, you have to register a future plan prior to calling the function. This code says to use a multisession plan to split the work across 4 cores, instead of the deafult of 1. You can look at the <a href="https://cran.r-project.org/web/packages/future/vignettes/future-1-overview.html"><code>future</code> documentation</a> to figure out what plan is best for what you’re doing. Note that only multisession works for RStudio, but multicore might work better for some purposes, such as working on a computing cluster. After registering a future plan, you can call <code><a href="../reference/get_clusters.html">get_clusters()</a></code> like normal and it should run faster!</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">doFuture</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/doFuture/man/registerDoFuture.html">registerDoFuture</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/future/man/plan.html">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/future/man/multisession.html">multisession</a></span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>

<span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_clusters.html">get_clusters</a></span><span class="op">(</span><span class="va">tr</span>, <span class="va">locs</span><span class="op">)</span></code></pre></div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>This vignette includes only some of the ways that the functions in <code>regentrans</code> can be used. Be creative and have fun with how use use these functions to probe pathogen transmission, but always remember to think about the specifics of your dataset and the questions you’re asking and what that means about interpreting your results!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Sophie Hoffman, Zena Lapp, Joyce Wang, Evan Snitkin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
