---
title: "Regentrans Vignette"
author: "Sophie Hoffman"
date: "3/29/2021"
output: 
  html_document:
    toc: true
  
---

The goal of `regentrans` is to give users a framework for using genomics to study regional pathogen transmission. 

To install the package from github, use the command

```{r, eval=FALSE, message = FALSE}
devtools::install_github('Snitkin-Lab-Umich/regentrans')
library(regentrans)
```

```{r,echo=FALSE, message=FALSE}
devtools::load_all()
```

Other packages you'll need to load to run this tutorial include:

```{r setup,echo=TRUE,message=FALSE}
library(ape)
library(dplyr)
library(devtools)
library(ggtree)
library(pheatmap)
library(ggplot2)
library(phytools)
library(gridExtra)

#set theme 
theme_set(theme_bw() + theme(strip.background = element_rect(fill="white",linetype='blank'), text=element_text(size=15)))
```

# Data Prep

In order to use the regentrans package, you will need your data in a certain format. It is important to ensure that you have all relevant information for all of the isolates of interest for your analysis. Each isolate should be represented in each metadata named list as well as in the fasta file of variants and phylogenetic tree. 

### Metadata

Each metadata category should appear as a named list where the names correspond to the isolate IDs and the items in the list reflect the category of the isolate. Locations for each isolate are required, while patient information, collection dates and sequence types are all optional inputs. 

To create named lists from a data.frame, you can do the following
```{r, eval = FALSE, message=FALSE}
locs <- metadata$locations
names(locs) <- metadata$isolate_ID
```

### Fasta file of variants

A fasta file of variants. Can be read in to R using the following command 

```{r, eval = FALSE, message=FALSE}
fasta <- ape::read.dna("/path/to/data.fasta",format = "fasta")

```

### SNV distance matrix

A pairwise SNV distance matrix. Can be created using the fasta file

```{r, eval = FALSE, message=FALSE}
dists <- ape::dist.dna(x = fasta, as.matrix = TRUE, model = "N", pairwise.deletion = TRUE)
```

### Phylogenetic tree

A phylogenetic tree (.treefile). Can be read in to R using the following command

```{r, eval = FALSE, message=FALSE}
tree <- ape::read.tree("path/to/data.treefile")
```

### Patient transfer network

A patient transfer network representing facilities that isolates came from represented by a data.frame with three columns: `source_facil`, `dest_facil` and `n_transfers`. 

### Input your data

If you want to use your own data as input to the functions in the vignette, format your data according to the instructions above and set the following objects equal to your data. If you will not be using any of the optional data, make sure to assign those objects to `NULL`. 

```{r, message=FALSE}
#named list of locations
locs <- regentrans::locs
#named list of patient IDs
pt <- regentrans::pt
#named list of isolate collection dates
dates <- regentrans::dates
#named list of sequence types
st <- NULL
#fasta file 
fasta <- regentrans::fasta
#SNV distance matrix 
dists <- regentrans::dists
#phylogenetic tree
tree <- regentrans::tr
#patient transfer network
pt_flow <- regentrans::pt_flow
```

### Subsetting to one isolate per patient per facility

If you have patient data for your isolates, it is a good idea to subset to one isolate from each patient per facility to avoid overestimating transmission. 

```{r}
isolate_subset <- cbind(locs, pt) %>% unique() %>% rownames()
locs <- locs[isolate_subset]
pt <- pt[isolate_subset]
dates <- dates[isolate_subset]
st <- st[isolate_subset]
```


# Is transmission occurring? 

In order to explore this question, we can leverage our WGS data in the form of the SNV distance matrix and the `get_snv_dists()` function. 


```{r, message=FALSE}
snv_dists <- get_snv_dists(dists = dists, locs = locs, pt = pt) 
```

In order to subset to one representation of each isolate-isolate pair (undirected), use the `subset_pairs()` function with the `snv_dists` object as input. 

```{r, message=FALSE}
snv_dists_sub <- subset_pairs(snv_dists)
ggplot() + 
  geom_histogram(data = subset(snv_dists_sub,Pair_Type == "Inter-facility pair"), mapping = aes(x = Pairwise_Dists, fill = "Inter-facility pair"), alpha = 0.2) + 
  geom_histogram(data = subset(snv_dists_sub,Pair_Type == "Intra-facility pair"), mapping = aes(x = Pairwise_Dists, fill = "Intra-facility pair"), alpha = 0.4) + labs(x = "Pairwise SNV Distance", y = "Count", fill = "Legend") +  ggtitle("SNV distances of intra- and inter- facility pairs")
```

- This plot can give you an idea of the population structure of your isolates. 

We can also take a closer look at the closely related pairs which can give us information about how transmission is occurring in the population.

```{r, message=FALSE}
ggplot() + 
  geom_histogram(data = subset(snv_dists_sub,Pair_Type == "Inter-facility pair" & Pairwise_Dists < 20), mapping = aes(x = Pairwise_Dists, fill = "Inter-facility pair"), alpha = 0.2, binwidth = 1) + 
  geom_histogram(data = subset(snv_dists_sub,Pair_Type == "Intra-facility pair" & Pairwise_Dists < 20), mapping = aes(x = Pairwise_Dists, fill = "Intra-facility pair"), alpha = 0.4, binwidth = 1) + labs(x = "Pairwise SNV Distance", y = "Count", fill = "Legend") +  ggtitle("SNV distances of intra- and inter- facility pairs \n< 20 SNV Distance")
```

# Is transmission occurring within and/or between locations?

In order to explore this question we can leverage our WGS data in the form of the SNV distance matrix using the `get_frac_intra()` function and the phylogenetic tree using the `get_clusters()` function. 

### get_frac_intra()

```{r, message=FALSE}
frac_intra <- get_frac_intra(snv_dists = snv_dists) #or get_frac_intra(dists = dists, locs = locs, pt = pt)

ggplot(data = frac_intra, aes(x = Thresh, y = Frac_Intra)) + geom_bar(stat = "identity", fill = "blue", alpha = 0.5) + labs(x = "SNV Distance Threshold", y = "Fraction of Intra-Facility Pairs") +  ggtitle("Fraction of Isolate Pairs from the Same \nFacility at Each SNV Distance Threshold") + ylim(0, 1) 
```

### get_clusters()

When working with the `get_clusters()` function, it is a good idea to perform analyses with various pureness levels. The default pureness value is 1. 

```{r, message=FALSE}
clusters <- get_clusters(tr,locs, pureness = 1)
#dissect output
pure_subtree_info <- clusters$pure_subtree_info
subtrees <- clusters$subtrees

ggplot(data = pure_subtree_info, aes(x = f_id, y = subtr_size)) + geom_jitter(position = position_jitter(width = 0.2, height = 0.1), alpha = 0.2, color = "blue") + labs(y = "Number of Isolates in \nPhylogenetic Cluster", x = "Facility ID", title = "Pure Subtree Information")
```

### Visualize the Phylogenetic Tree Annotated by Location
```{r, message=FALSE}
tr_sub <- ape::keep.tip(tr, names(locs))
mpr_tree = midpoint.root(tr_sub)
gheatmap(ggtree(mpr_tree, layout = "circular"),as.data.frame(locs), color = NA)
```

# What locations is transmission occurring within/between?

### SNV distances 

Choose and SNV cutoff (it is recommended to perform a sensitivity analysis by choosing several different SNV thresholds and seeing how robust the results are to these changes). 

- One way to find an initial SNV cutoff for your data is to look at the `Fraction of Isolate Pairs from the Same Facility at Each SNV Distance Threshold` plot where the fraction of intra-facility pairs drops off. *Note:* Not all datasets will drop off, use this method only if your plot has a similar shape to the vignette data.

- Other suggestions about how to choose a SNV cutoff can be found in the regentrans manuscript. 

Using this cutoff, the number of closely related pairs within and between facilities can be determined and used to identify facilities and facility pairs with more putative spread. 


```{r, message=FALSE}
cutoff <- 10 #adjust according to your own analysis!
snv_dists %>% filter(Pairwise_Dists <= cutoff) %>% group_by(Loc1, Pair_Type) %>% summarize(n = n()) %>% ggplot(aes(x = Loc1, y = n, fill = Pair_Type)) + geom_bar(position = "dodge", stat = "identity") + labs(y = "Number of Closely Related Isolate Pairs", x = "Facility ID", title = "Number of Closely Related Isolate Pairs \nby Facility")

```


```{r, message=FALSE}
snv_dists %>% filter(Pairwise_Dists <= cutoff) %>% group_by(Loc1, Loc2) %>% summarize(n = n()) %>% filter(Loc1 != Loc2) %>% subset_pairs() %>% mutate(Facil_Pair = paste(Loc1, Loc2)) %>% filter(n >=  mean(n)) %>% ggplot(aes(x = Facil_Pair, y = n)) + geom_bar(position = "dodge", stat = "identity", fill = "salmon") + labs(y = "Number of Closely Related Isolate Pairs", x = "Facility Pair", title = "Number of Closely Related Isolate Pairs \nby Facility Pair") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

### Permutation

Permutation can be used as a method to check whether your data is enriched for transmission within/between certain facilities. For example, if we permute the location labels on the `snv_dists` output before calculating the number of isolates below a certain threshold at each facility, we can interrogate whether the enrichment in our data are significant. Permutation can also be used to address significance of other regentrans methods. 

```{r, message=FALSE, warning=FALSE}
permut <- setNames(as.data.frame(cbind(rep(unique(locs), each = 2), c("Inter-facility pair", "Intra-facility pair"))), c("Loc1_mut", "Pair_Type"))

for(i in 1:100){
  #for reproducibility
  set.seed(i)
  summary  <- snv_dists %>% mutate(Loc1_mut = sample(Loc1)) %>% filter(Pairwise_Dists <= cutoff) %>% group_by(Loc1_mut, Pair_Type) %>% summarize(n = n())
  permut <- left_join(permut, summary, by = c("Loc1_mut" = "Loc1_mut", "Pair_Type" = "Pair_Type"))
}

permut %>% `colnames<-`(c("Loc1", "Pair_Type", 1:100)) %>% tidyr::pivot_longer(!c(Loc1, Pair_Type), names_to = "run", values_to = "value") %>% ggplot() + geom_boxplot(mapping = aes(x = Loc1, y = value, fill = Pair_Type)) + geom_point(data = (snv_dists %>% filter(Pairwise_Dists <= cutoff) %>% group_by(Loc1, Pair_Type) %>% summarize(n = n())), mapping = aes(x = Loc1, y = n, color = Pair_Type), shape = 15) + scale_color_manual(values=c("red", "blue")) + labs(y = "Number of Closely Related \nIsolate Pairs (permuted)", x = "Facility ID", title = "Permuted Number of Closely Related Isolate Pairs \nby Facility with real-value points")
```

This plot suggests that facility D is enriched for intra-facility transmission. 

### Fsp

Fsp is a measure of similarity between populations at different facilities. Value of 0 indicates that the populations are identical, while value of 1 indicates that the populations are completely different. 

```{r, eval=FALSE, message=FALSE}
Fsp <- get_facility_fsp(fasta, locs)
pheatmap(Fsp)
```

```{r, echo=FALSE, message=FALSE}
Fsp <- regentrans::Fsp
pheatmap(Fsp)
```

One way to determine whether Fsp values are significant is to use permutation (Donker et al. 2017)

### Is transmission occurring between different sample sources?

If your dataset includes both environmental samples and patient samples, you could utilize the `pt` input to the `get_snv_dists()` function to explore whether transmission is occurring between different sample sources by considering each environmental source a "patient". You could then determine what fraction of your isolate pairs are patient-patient pairs vs. patient-environment pairs vs. environment-environment pairs with the code below. 

- This will be a **toy example** as this sample data only includes patient samples. For this example case we will say that any patient ID > 300 is an environmental source. 

- If your samples are labeled differently (i.e. patient samples are labeled with PT_... and environmental samples are labeled with EN_...) you may want to use a function like `grepl` to differentiate between sample types. 

- It is important to consider whether you have subsetted to one sample per "patient" (or sample) source per location before performing this analysis. 

```{r, message = FALSE}
thresh = 10 

snv_dists %>% filter(Pair_Type == "Intra-facility pair") %>% mutate(Pt_Pair_Type = ifelse(Patient1 < 300 & Patient2 < 300, "pt-pt", ifelse(Patient1 > 300 & Patient2 > 300, "env-env", "pt-env"))) %>% filter(Pairwise_Dists <= thresh) %>% group_by(Pt_Pair_Type) %>% summarize(n = n()) %>% ggplot() + geom_bar(mapping = aes(x = Pt_Pair_Type, y = n), stat = "Identity", fill = "salmon") + labs(y = "Number of Closely Related Isolate Pairs", x = "Sample Pair Type", title = "Number of Closely Related Isolate Pairs \nby Sample Pair Type")
```

# Have transmission dynamics changed over time or across species?

In order to investigate whether transmission dynamics have changed over time or differ across groups (such as sequence type or species) we can "facet" on those variables. Here is an example of faceting on date:

```{r, message=FALSE}
snv_dists_sub_dates <- snv_dists_sub %>% left_join(bind_cols(names(dates), dates), by = c("Isolate1" = "...1")) %>% rename("date" = "...2") 

ggplot() + 
  geom_histogram(data = subset(snv_dists_sub_dates,Pair_Type == "Inter-facility pair"), mapping = aes(x = Pairwise_Dists, fill = "Inter-facility pair"), alpha = 0.2) + 
  geom_histogram(data = subset(snv_dists_sub_dates,Pair_Type == "Intra-facility pair"), mapping = aes(x = Pairwise_Dists, fill = "Intra-facility pair"), alpha = 0.4) + labs(x = "Pairwise SNV Distance", y = "Count", fill = "Legend") +  ggtitle("SNV distances of intra- and inter- facility pairs \nby Date of Isolate Collection") + facet_wrap(~date)
```

# Is transmission occurring along paths of higher patient flow? 

In order to explore this question, we can leverage our WGS data in the form of the SNV distance matrix along with the patient flow network using the `patient_transfer()` function.

*Note:* If you do not have a patient transfer network, you could use geographic distances between facilities as a proxy for patient transfer because they have been shown to correlate. 

**Choosing a pairwise SNV distance threshold based on your dataset is very important, avoid using the default value**

```{r, message=FALSE}
pt_trans <- patient_transfer(pt_trans_net = pt_flow, snv_dists = snv_dists, thresh = 10) %>% dplyr::mutate(avg_direct_transfers = rowMeans(select(., dplyr::ends_with("transfers")), na.rm = TRUE), avg_indirect_flow = rowMeans(select(., dplyr::starts_with("indirect")), na.rm = TRUE)) 

ggplot(data = pt_trans, mapping = aes(x = avg_direct_transfers, y = n_closely_related_pairs)) + geom_jitter(position = position_jitter(width = 0.2, height = 0.1), alpha = 0.2, color = "blue") + labs(y = "Number of Closely Related Isolate Pairs", x = "Number of Transfers", title = "Number of Patient Transfers between Facilities \nvs. Number of Closely Related Isolate Pairs at Facilities \nfor Each Facility Pair") + geom_smooth(method = "lm")

```

*Note:* This data is simulated

```{r, message=FALSE}
Fsp %>% long_form() %>% subset_pairs() %>% left_join(pt_trans, by = c("Facil_1" = "Loc1", "Facil_2" = "Loc2")) %>% ggplot(mapping = aes(x = avg_direct_transfers, y = Fsp_val)) + geom_jitter(position = position_jitter(width = 0.2, height = 0.1), alpha = 0.2, color = "blue") + labs(y = "Fsp (population-level similarity)", x = "Number of Transfers", title = "Number of Patient Transfers between Facilities \nvs. Fsp (population-level similarity) between Facilities \nfor Each Facility Pair") + geom_smooth(method = "lm")

```

Fsp values closer to 0 indicate more closely related populations.

# Are there any observable geographic trends in prevalence/transmission?

In order to evaluate whether there are any observable geographic trends in prevalence or transmission, we will need geographic information (latitude and longitude) for each facility. The data.frame should look like this, and be called `facil_coord`

```{r, echo = FALSE, message=FALSE}
cities <- force(us.cities) %>% filter(!country.etc %in% c("HI", "AK"))
facilities <- unique(locs)
cities_sub <- cities[1:length(facilities),4:5]
facil_coord <- as.data.frame(cbind(facilities, cities_sub))
head(facil_coord)
```
Previous studies have found that there is a strong relationship between extent of patient transfer and geographic distance (Wang J et al. 2020). Thus, geographic distance can be used as a proxy for patient transfer if it is not available. 

We can combine this data with our `locs` data, `Fsp` data and `pt_trans` network to create a geographic map where point size represents number of isolates collected from a facility and line width represents Fsp value (population similarity between facilities).

```{r, message=FALSE}
loc_n <- locs %>% table() %>% as.data.frame() %>% `colnames<-`(c("facilities", "n"))

facil_geo <- facil_coord %>% left_join(loc_n)

Fsp_geo <- Fsp %>% long_form() %>% subset_pairs() %>% left_join(facil_geo, by = c("Facil_1" = "facilities")) %>% rename(facil_1_lat = lat, facil_1_long = long, facil_1_n = n) %>% left_join(facil_geo, by = c("Facil_2" = "facilities")) %>% rename(facil_2_lat = lat, facil_2_long = long, facil_2_n = n)

facil_geo %>% ggplot(aes(x=long, y=lat)) +
  geom_point(aes(size=n), alpha=0.5, color = "blue") + 
  borders("state") +
  coord_quickmap() +
  geom_curve(aes(x = facil_1_long, y = facil_1_lat, xend = facil_2_long, yend = facil_2_lat,
                 alpha = (1/Fsp_val)),
             data = Fsp_geo, curvature = 0.33) +
  theme_void() + theme(legend.position = "none") 


```

*Note:* This location data is simulated

Another way to quantify closely related populations is by the number of isolate pairs between them that are below a SNV distance threshold. 

**Choosing a pairwise SNV distance threshold based on your dataset is very important, avoid using the default value**

```{r, message=FALSE}
thresh = 10

snv_dist_summary <- snv_dists %>% dplyr::group_by(Loc1, Loc2) %>% dplyr::summarize(n_closely_related_pairs = sum(Pairwise_Dists <= thresh)) %>% dplyr::filter(Loc1 != Loc2)

snv_geo <- snv_dists %>% dplyr::group_by(Loc1, Loc2) %>% dplyr::summarize(n_closely_related_pairs = sum(Pairwise_Dists <= thresh)) %>% dplyr::filter(Loc1 != Loc2) %>% left_join(facil_geo, by = c("Loc1" = "facilities")) %>% rename(facil_1_lat = lat, facil_1_long = long, facil_1_n = n) %>% left_join(facil_geo, by = c("Loc2" = "facilities")) %>% rename(facil_2_lat = lat, facil_2_long = long, facil_2_n = n)

facil_geo %>% ggplot(aes(x=long, y=lat)) +
  geom_point(aes(size=n), alpha=0.5, color = "blue") + 
  borders("state") +
  coord_quickmap() +
  geom_curve(aes(x = facil_1_long, y = facil_1_lat, xend = facil_2_long, yend = facil_2_lat,
                 alpha = n_closely_related_pairs),
             data = snv_geo, curvature = 0.33) +
  theme_void() + theme(legend.position = "none") 

```

*Note:* This location data is simulated

These are just two examples of how to plot the geographic, but there may be other proxies for patient transfer that work better for your analysis. 

